<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Post Hoc Multiple Testing Inference • sanssouci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Post Hoc Multiple Testing Inference">
<meta property="og:description" content="Post hoc inference based on the control of a dedicated risk called the Joint (Family-Wise) Error Rate.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sanssouci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a>
    </li>
    <li>
      <a href="articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a>
    </li>
    <li>
      <a href="articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a>
    </li>
    <li>
      <a href="articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a>
    </li>
    <li>
      <a href="articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a>
    </li>
    <li>
      <a href="articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a>
    </li>
    <li>
      <a href="articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/sanssouci/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="sanssouci">sanssouci<a class="anchor" aria-label="anchor" href="#sanssouci"></a>
</h1></div>
<!-- badges: start -->
<p>The goal of sanssouci [sãsusi] is to perform <strong>post hoc inference</strong>: in a multiple testing context, sanssouci provides statistical guarantees on possibly user-defined and/or data-driven sets of hypotheses.</p>
<p>Typical use cases include:</p>
<ul>
<li>
<strong>Differential gene expression (DGE) studies in genomics</strong>:<ul>
<li>see vignettes dedicated to <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression.html">microarray data</a> and <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression_RNAseq.html">RNAseq data</a>
</li>
<li>run <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">volcano plot shiny app</a>
</li>
</ul>
</li>
<li>
<strong>fMRI studies in neuroimaging</strong>: see <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_fMRI.html">fMRI vignette</a>.</li>
</ul>
<p>In both cases, the permutation-based post hoc inference methods implemented in the <code>sanssouci</code> package outperform classical post hoc bounds based on probabilistic inequalities.</p>
<div class="section level2">
<h2 id="example-differential-analysis">Example: differential analysis<a class="anchor" aria-label="anchor" href="#example-differential-analysis"></a>
</h2>
<p>We briefly illustrate the case of differential gene expression study (see dedicated <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression.html">vignette</a> for more information). We start by creating an object of class <code>sanssouci</code> from gene expression data available from the <a href="https://github.com/pneuvial/sanssouci.data" class="external-link">sanssouci.data</a> package,</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/pneuvial/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; remotes::install_github("pneuvial/sanssouci.data")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">expr_ALL</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">expr_ALL</span><span class="op">)</span> <span class="op">==</span> <span class="st">"BCR/ABL"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span> <span class="co"># map to 0/1</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">expr_ALL</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span></span></code></pre></div>
<p>Then we fit the method by with <code>B=1000</code> permutations, and with a target risk <code>alpha = 0.1</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="fl">0.1</span>, B <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="output-1---post-hoc-bound-for-a-subset-of-genes">Output 1 - Post hoc bound for a subset of genes<a class="anchor" aria-label="anchor" href="#output-1---post-hoc-bound-for-a-subset-of-genes"></a>
</h3>
<p>Let us assume that we are interested in genes with a (p)-value below (10^{-3}).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="reference/all-generics.html">pValues</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">1e-3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: namespace 'sansSouci' is not available and has been replaced</span></span>
<span><span class="co">#&gt; by .GlobalEnv when processing object '&lt;unknown&gt;'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">S</span><span class="op">)</span></span>
<span><span class="co">#&gt;          TP         FDP </span></span>
<span><span class="co">#&gt; 114.0000000   0.2919255</span></span></code></pre></div>
<p>The method ensures with (1-= 90%) confidence that there are at least 114 truly differentially expressed genes (true positives, TP) among these 161 genes, corresponding to a false discovery proportion (FDP) less than 0.3.</p>
</div>
<div class="section level3">
<h3 id="output-2---confidence-curves-for-top-k-feature-lists">Output 2 - Confidence curves for “top-k” feature lists<a class="anchor" aria-label="anchor" href="#output-2---confidence-curves-for-top-k-feature-lists"></a>
</h3>
<p>Another output of the method is the following (1-)-level confidence curve on FDP (left panel) and corresponding TP (right panel) in “top-k” gene lists.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span>, xmax <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">S</span><span class="op">)</span>, </span>
<span>             color <span class="op">=</span> <span class="st">"gray"</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-conf-curve-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="output-3---volcano-plots">Output 3 - Volcano plots<a class="anchor" aria-label="anchor" href="#output-3---volcano-plots"></a>
</h3>
<p>A more refined user selection is obtained by selecting genes based on both significance (here, a FDR-adjusted (p)-value below 0.05) and fold change (here, fold change above 0.3). Such selections are generally represented by volcano plots:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, q <span class="op">=</span> <span class="fl">0.05</span>, r <span class="op">=</span> <span class="fl">0.3</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-volcano-plot-1.png" width="100%"></p>
<p>Importantly, multiple such selections can be made without compromising the validity of the bounds. The <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">IIDEA shiny app</a> makes it possible to perform such selections interactively and without programming.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<!-- You can install the released version of sanssouci from [CRAN](https://CRAN.R-project.org) with: --><!-- ``` r --><!-- install.packages("sanssouci") --><!-- ``` --><p>You can install the development version from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pneuvial/sanssouci"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/pneuvial/sanssouci/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/pneuvial/sanssouci/issues" class="external-link">Report a bug</a></li>
<li><a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">Run volcano plot shiny app</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sanssouci</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.math.univ-toulouse.fr/~pneuvial/" class="external-link">Pierre Neuvial</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3584-9998" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://www.imo.universite-paris-saclay.fr/~blanchard/" class="external-link">Gilles Blanchard</a> <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://durandg12.github.io/" class="external-link">Guillermo Durand</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4056-5631" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Nicolas Enjalbert-Courrech <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://etienneroquain-81.webself.net/" class="external-link">Etienne Roquain</a> <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/pneuvial/sanssouci/actions" class="external-link"><img src="https://github.com/pneuvial/sanssouci/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/github/pneuvial/sanssouci/branch/develop" class="external-link"><img src="https://codecov.io/gh/pneuvial/sanssouci/branch/develop/graph/badge.svg" alt="Coverage Status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.math.univ-toulouse.fr/~pneuvial/" class="external-link">Pierre Neuvial</a>, <a href="https://www.imo.universite-paris-saclay.fr/~blanchard/" class="external-link">Gilles Blanchard</a>, <a href="https://durandg12.github.io/" class="external-link">Guillermo Durand</a>, Nicolas Enjalbert-Courrech, <a href="https://etienneroquain-81.webself.net/" class="external-link">Etienne Roquain</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
