<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Post Hoc Multiple Testing Inference • sanssouci</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Post Hoc Multiple Testing Inference">
<meta name="description" content="Post hoc inference based on the control of a dedicated risk called the Joint (Family-Wise) Error Rate.">
<meta property="og:description" content="Post hoc inference based on the control of a dedicated risk called the Joint (Family-Wise) Error Rate.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.16.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="articles/post-hoc_differential-expression_linear-model.html">Bootstrap-based post hoc inference for differential gene expression studies in linear models</a></li>
    <li><a class="dropdown-item" href="articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for two-group differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for two-group differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="sanssouci">sanssouci<a class="anchor" aria-label="anchor" href="#sanssouci"></a>
</h1></div>
<!-- badges: start -->

<p>The goal of sanssouci [sãsusi] is to perform <strong>post hoc inference</strong>: in a multiple testing context, sanssouci provides statistical guarantees on possibly user-defined and/or data-driven sets of hypotheses.</p>
<p>Typical use cases include:</p>
<ul>
<li>
<strong>Differential gene expression (DGE) studies in genomics</strong>:
<ul>
<li>see vignettes dedicated to <a href="articles/post-hoc_differential-expression.html">microarray data</a> and <a href="articles/post-hoc_differential-expression_RNAseq.html">RNAseq data</a>
</li>
<li>run <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">volcano plot shiny app</a>
</li>
</ul>
</li>
<li>
<strong>fMRI studies in neuroimaging</strong>: see <a href="articles/post-hoc_fMRI.html">fMRI vignette</a>.</li>
</ul>
<p>In both cases, the permutation-based post hoc inference methods implemented in the <code>sanssouci</code> package outperform classical post hoc bounds based on probabilistic inequalities.</p>
<div class="section level2">
<h2 id="example-differential-analysis">Example: differential analysis<a class="anchor" aria-label="anchor" href="#example-differential-analysis"></a>
</h2>
<p>We briefly illustrate the case of differential gene expression study (see dedicated <a href="articles/post-hoc_differential-expression.html">vignette</a> for more information). We start by creating an object of class <code>sanssouci</code> from gene expression data available from the <a href="https://github.com/sanssouci-rog/sanssouci.data" class="external-link">sanssouci.data</a> package,</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; remotes::install_github("sanssouci-org/sanssouci.data")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">expr_ALL</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">expr_ALL</span><span class="op">)</span> <span class="op">==</span> <span class="st">"BCR/ABL"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span> <span class="co"># map to 0/1</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">expr_ALL</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span></span></code></pre></div>
<p>Then we fit the method by with <code>B=1000</code> permutations, and with a target risk <code>alpha = 0.1</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="fl">0.1</span>, B <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="output-1---post-hoc-bound-for-a-subset-of-genes">Output 1 - Post hoc bound for a subset of genes<a class="anchor" aria-label="anchor" href="#output-1---post-hoc-bound-for-a-subset-of-genes"></a>
</h3>
<p>Let us assume that we are interested in genes with a <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value below <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>10</mn><mrow><mo>−</mo><mn>3</mn></mrow></msup><annotation encoding="application/x-tex">10^{-3}</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="reference/all-generics.html">pValues</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">1e-3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res</span>, <span class="va">S</span><span class="op">)</span></span>
<span><span class="co">#&gt;          TP         FDP </span></span>
<span><span class="co">#&gt; 118.0000000   0.2670807</span></span></code></pre></div>
<p>The method ensures with <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>α</mi><mo>=</mo><mn>90</mn><mi>%</mi></mrow><annotation encoding="application/x-tex">1-\alpha = 90\%</annotation></semantics></math> confidence that there are at least 118 truly differentially expressed genes (true positives, TP) among these 161 genes, corresponding to a false discovery proportion (FDP) less than 0.27.</p>
</div>
<div class="section level3">
<h3 id="output-2---confidence-curves-for-top-k-feature-lists">Output 2 - Confidence curves for “top-k” feature lists<a class="anchor" aria-label="anchor" href="#output-2---confidence-curves-for-top-k-feature-lists"></a>
</h3>
<p>Another output of the method is the following <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>α</mi></mrow><annotation encoding="application/x-tex">1-\alpha</annotation></semantics></math>-level confidence curve on FDP (left panel) and corresponding TP (right panel) in “top-k” gene lists.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span>, xmax <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">S</span><span class="op">)</span>, </span>
<span>             color <span class="op">=</span> <span class="st">"gray"</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; ℹ Please use `linewidth` instead.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span></code></pre></div>
<p><img src="reference/figures/README-conf-curve-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="output-3---volcano-plots">Output 3 - Volcano plots<a class="anchor" aria-label="anchor" href="#output-3---volcano-plots"></a>
</h3>
<p>A more refined user selection is obtained by selecting genes based on both significance (here, a FDR-adjusted <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value below 0.05) and fold change (here, fold change above 0.3). Such selections are generally represented by volcano plots:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, q <span class="op">=</span> <span class="fl">0.05</span>, r <span class="op">=</span> <span class="fl">0.3</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-volcano-plot-1.png" width="100%"></p>
<p>Importantly, multiple such selections can be made without compromising the validity of the bounds. The <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">IIDEA shiny app</a> makes it possible to perform such selections interactively and without programming.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<!-- You can install the released version of sanssouci from [CRAN](https://CRAN.R-project.org) with: -->
<p>You can install the development version from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("pak")</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"sanssouci-org/sanssouci"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a>
</h2>
<p>This project uses <a href="https://semver.org/" class="external-link">Semantic Versioning</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sanssouci-org/sanssouci/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/sanssouci-org/sanssouci/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sanssouci</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Pierre Neuvial <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3584-9998" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Gilles Blanchard <br><small class="roles"> Author </small>   </li>
<li>Guillermo Durand <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4056-5631" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Nicolas Enjalbert-Courrech <br><small class="roles"> Author </small>   </li>
<li>Etienne Roquain <br><small class="roles"> Author </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sanssouci-org/sanssouci/actions" class="external-link"><img src="https://github.com/sanssouci-org/sanssouci/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/github/sanssouci-org/sanssouci/branch/develop" class="external-link"><img src="https://codecov.io/gh/sanssouci-org/sanssouci/branch/develop/graph/badge.svg" alt="Coverage Status"></a></li>
<li><a href="https://github.com/sanssouci-org/sanssouci/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/sanssouci-org/sanssouci/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
