<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ sanssouci</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.16.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_linear-model.html">Bootstrap-based post hoc inference for differential gene expression studies in linear models</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for two-group differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for two-group differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="SansSouci-methods.html">nHyp(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">nObs(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">label(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">print(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">pValues(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">foldChanges(<i>&lt;SansSouci&gt;</i>)</a></code> <code><a href="SansSouci-methods.html">thresholds(<i>&lt;SansSouci&gt;</i>)</a></code>

        </dt>
        <dd>Basic methods for class <code>SansSouci</code></dd>
      </dl><dl><dt>

          <code><a href="SansSouci.html">SansSouci()</a></code>

        </dt>
        <dd>Create an object of class 'SansSouci'</dd>
      </dl><dl><dt>

          <code><a href="SansSouciDyadic.html">SansSouciDyadic()</a></code>

        </dt>
        <dd>Create an object of class 'SansSouciStruct' with a dyadic structure</dd>
      </dl><dl><dt>

          <code><a href="SansSouciSim.html">SansSouciSim()</a></code>

        </dt>
        <dd>Create an object of class 'SansSouci' from simulations</dd>
      </dl><dl><dt>

          <code><a href="SansSouciStruct-methods.html">nHyp(<i>&lt;SansSouciStruct&gt;</i>)</a></code> <code><a href="SansSouciStruct-methods.html">nLeaves(<i>&lt;SansSouciStruct&gt;</i>)</a></code> <code><a href="SansSouciStruct-methods.html">label(<i>&lt;SansSouciStruct&gt;</i>)</a></code> <code><a href="SansSouciStruct-methods.html">pValues(<i>&lt;SansSouciStruct&gt;</i>)</a></code>

        </dt>
        <dd>Basic methods for class <code>SansSouciStruct</code></dd>
      </dl><dl><dt>

          <code><a href="SansSouciStruct.html">SansSouciStruct()</a></code>

        </dt>
        <dd>Create an object of class 'SansSouciStruct'</dd>
      </dl><dl><dt>

          <code><a href="V.star.html">V.star()</a></code>

        </dt>
        <dd>Post hoc bound on the number of false positives</dd>
      </dl><dl><dt>

          <code><a href="all-generics.html">nHyp()</a></code> <code><a href="all-generics.html">nObs()</a></code> <code><a href="all-generics.html">label()</a></code> <code><a href="all-generics.html">pValues()</a></code> <code><a href="all-generics.html">foldChanges()</a></code> <code><a href="all-generics.html">thresholds()</a></code> <code><a href="all-generics.html">nLeaves()</a></code>

        </dt>
        <dd>Generic functions for S3 class SansSouci</dd>
      </dl><dl><dt>

          <code><a href="bootstrap_permutation.html">bootstrap_permutation()</a></code>

        </dt>
        <dd>Mass-univariate bootstrap-based inference for contrasts in a linear model</dd>
      </dl><dl><dt>

          <code><a href="calibrate.html">calibrate()</a></code> <code><a href="calibrate.html">calibrate0()</a></code>

        </dt>
        <dd>Perform JER calibration from randomization p-values</dd>
      </dl><dl><dt>

          <code><a href="calibration_bootstap.html">calibration_bootstap()</a></code>

        </dt>
        <dd>Calibration of post hoc bound using bootstrap permutations</dd>
      </dl><dl><dt>

          <code><a href="confCurveFromFam.html">confCurveFromFam()</a></code>

        </dt>
        <dd>Confidence bounds on the true/false positives among most significant items</dd>
      </dl><dl><dt>

          <code><a href="curve.V.star.forest.html">curve.V.star.forest.naive()</a></code> <code><a href="curve.V.star.forest.html">curve.V.star.forest.fast()</a></code>

        </dt>
        <dd>Compute a curve of post hoc bounds based on a reference family with forest structure</dd>
      </dl><dl><dt>

          <code><a href="curveMaxFP.html">curveMaxFP()</a></code>

        </dt>
        <dd>Upper bound for the number of false discoveries among most significant items</dd>
      </dl><dl><dt>

          <code><a href="curveMaxFP_alternatives.html">curveMaxFP_BNR2014()</a></code> <code><a href="curveMaxFP_alternatives.html">curveMaxFP_Mein2006()</a></code>

        </dt>
        <dd>Upper bound for the number of false discoveries among most significant items</dd>
      </dl><dl><dt>

          <code><a href="delete.gaps.html">delete.gaps()</a></code>

        </dt>
        <dd>Delete the gaps induced by pruning</dd>
      </dl><dl><dt>

          <code><a href="dyadic.html">dyadic.from.leaf_list()</a></code> <code><a href="dyadic.html">dyadic.from.window.size()</a></code> <code><a href="dyadic.html">dyadic.from.height()</a></code>

        </dt>
        <dd>Create a complete dyadic tree structure</dd>
      </dl><dl><dt>

          <code><a href="fit.SansSouci.html">fit(<i>&lt;SansSouci&gt;</i>)</a></code>

        </dt>
        <dd>Fit SansSouci object</dd>
      </dl><dl><dt>

          <code><a href="fit.SansSouciStruct.html">fit(<i>&lt;SansSouciStruct&gt;</i>)</a></code>

        </dt>
        <dd>Fit SansSouciStruct object</dd>
      </dl><dl><dt>

          <code><a href="forest.completion.html">forest.completion()</a></code>

        </dt>
        <dd>Complete a forest structure</dd>
      </dl><dl><dt>

          <code><a href="formatBounds.html">formatBounds()</a></code>

        </dt>
        <dd>Table of bounds</dd>
      </dl><dl><dt>

          <code><a href="gauss_bloc.html">gauss_bloc()</a></code>

        </dt>
        <dd>Generate a block of deterministic signal of Gaussian shape</dd>
      </dl><dl><dt>

          <code><a href="gaussianSamples.html">gaussianSamples()</a></code>

        </dt>
        <dd>Simulate equi-correlated data</dd>
      </dl><dl><dt>

          <code><a href="gaussianTestStatistics.html">gaussianTestStatistics()</a></code>

        </dt>
        <dd>Simulate Gaussian test statistics</dd>
      </dl><dl><dt>

          <code><a href="gen.mu.leaves.html">gen.mu.leaves()</a></code>

        </dt>
        <dd>Generate the signal in the leaves of the tree</dd>
      </dl><dl><dt>

          <code><a href="gen.mu.noleaves.html">gen.mu.noleaves()</a></code>

        </dt>
        <dd>Generate an unstructured signal</dd>
      </dl><dl><dt>

          <code><a href="gen.p.values.html">gen.p.values()</a></code>

        </dt>
        <dd>Generate one-sided p-values associated to a given signal with equi-correlated noise</dd>
      </dl><dl><dt>

          <code><a href="get_perm.html">get_perm()</a></code>

        </dt>
        <dd>Get permutation statistics and p-values</dd>
      </dl><dl><dt>

          <code><a href="get_pivotal_stat.html">get_pivotal_stat()</a></code>

        </dt>
        <dd>Get a vector of pivotal statistics associated to permutation p-values and to a reference family</dd>
      </dl><dl><dt>

          <code><a href="lm_test.html">lm_test()</a></code>

        </dt>
        <dd>Mass-univariate inference for contrasts in a linear model</dd>
      </dl><dl><dt>

          <code><a href="maxFDP.html">maxFDP()</a></code>

        </dt>
        <dd>Upper bound for the false discovery proportion in a selection</dd>
      </dl><dl><dt>

          <code><a href="maxFP.html">maxFP()</a></code>

        </dt>
        <dd>Upper bound for the number of false discoveries in a selection</dd>
      </dl><dl><dt>

          <code><a href="minTDP.html">minTDP()</a></code>

        </dt>
        <dd>Lower bound for the true discovery proportion in a selection</dd>
      </dl><dl><dt>

          <code><a href="minTP.html">minTP()</a></code>

        </dt>
        <dd>Lower bound for the number of true discoveries in a selection</dd>
      </dl><dl><dt>

          <code><a href="mini_batch_rowTestFUN.html">mini_batch_rowTestFUN()</a></code>

        </dt>
        <dd>Mini batch version of a rowTestFUN function.</dd>
      </dl><dl><dt>

          <code><a href="nb.elements.html">nb.elements()</a></code>

        </dt>
        <dd>Number of unique regions in a reference family with forest structure</dd>
      </dl><dl><dt>

          <code><a href="plot.SansSouci.html">plot(<i>&lt;SansSouci&gt;</i>)</a></code>

        </dt>
        <dd>Plot confidence bound on the true/false positives among most significant items</dd>
      </dl><dl><dt>

          <code><a href="plotConfCurve.html">plotConfCurve()</a></code>

        </dt>
        <dd>Plot confidence bound</dd>
      </dl><dl><dt>

          <code><a href="posthocBySimes.html">posthocBySimes()</a></code>

        </dt>
        <dd>post hoc bound obtained from Simes' inequality</dd>
      </dl><dl><dt>

          <code><a href="predict.SansSouci.html">predict(<i>&lt;SansSouci&gt;</i>)</a></code>

        </dt>
        <dd>Post hoc confidence bounds on the true/false positives</dd>
      </dl><dl><dt>

          <code><a href="pruning.html">pruning()</a></code>

        </dt>
        <dd>Prune a forest structure to speed up computations</dd>
      </dl><dl><dt>

          <code><a href="reference-families.html">t_inv_linear()</a></code> <code><a href="reference-families.html">t_linear()</a></code> <code><a href="reference-families.html">t_inv_beta()</a></code> <code><a href="reference-families.html">t_beta()</a></code>

        </dt>
        <dd>Reference families</dd>
      </dl><dl><dt>

          <code><a href="rowBinomialTests.html">rowBinomialTests()</a></code>

        </dt>
        <dd>Binomial proportion tests for each row of a matrix</dd>
      </dl><dl><dt>

          <code><a href="rowPearsonCorrelationTests.html">rowPearsonCorrelationTests()</a></code>

        </dt>
        <dd>Pearson's correlation test for rows of a matrix</dd>
      </dl><dl><dt>

          <code><a href="rowWelchTests.html">rowWelchTests()</a></code>

        </dt>
        <dd>Welch T-tests for rows of a matrix</dd>
      </dl><dl><dt>

          <code><a href="rowWilcoxonTests.html">rowWilcoxonTests()</a></code>

        </dt>
        <dd>Wilcoxon rank sum tests for each row of a matrix</dd>
      </dl><dl><dt>

          <code><a href="rowZTests.html">rowZTests()</a></code>

        </dt>
        <dd>Z tests for rows of a matrix</dd>
      </dl><dl><dt>

          <code><a href="row_lm_test.html">row_lm_test()</a></code>

        </dt>
        <dd>Mass-univariate bootstrap-based inference for contrasts in a linear model</dd>
      </dl><dl><dt>

          <code><a href="testByRandomization.html">testByRandomization()</a></code>

        </dt>
        <dd>Randomization-based testing</dd>
      </dl><dl><dt>

          <code><a href="volcanoPlot.html">volcanoPlot()</a></code>

        </dt>
        <dd>Volcano plot</dd>
      </dl><dl><dt>

          <code><a href="volcanoPlot.numeric.html">volcanoPlot(<i>&lt;numeric&gt;</i>)</a></code>

        </dt>
        <dd>Volcano plot</dd>
      </dl><dl><dt>

          <code><a href="zeta.html">zeta.HB()</a></code> <code><a href="zeta.html">zeta.kBonf()</a></code> <code><a href="zeta.html">zeta.trivial()</a></code> <code><a href="zeta.html">zeta.DKWM()</a></code>

        </dt>
        <dd>Estimate the number of true null hypotheses among a set of p-values</dd>
      </dl><dl><dt>

          <code><a href="zetas.tree.html">zetas.tree()</a></code>

        </dt>
        <dd>Estimate of the proportion of true nulls in each node of a tree</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

