<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Prune a forest structure to speed up computations — pruning • sanssouci</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prune a forest structure to speed up computations — pruning"><meta name="description" content="The pruned forest structure makes the computation of V.star(),
curve.V.star.forest.naive() and curve.V.star.forest.fast() faster."><meta property="og:description" content="The pruned forest structure makes the computation of V.star(),
curve.V.star.forest.naive() and curve.V.star.forest.fast() faster."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Prune a forest structure to speed up computations</h1>
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/v0.15.0/R/tree-functions.R" class="external-link"><code>R/tree-functions.R</code></a></small>
      <div class="d-none name"><code>pruning.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The pruned forest structure makes the computation of <code><a href="V.star.html">V.star()</a></code>,
<code><a href="curve.V.star.forest.html">curve.V.star.forest.naive()</a></code> and <code><a href="curve.V.star.forest.html">curve.V.star.forest.fast()</a></code> faster.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pruning</span><span class="op">(</span><span class="va">C</span>, <span class="va">ZL</span>, <span class="va">leaf_list</span>, prune.leafs <span class="op">=</span> <span class="cn">FALSE</span>, delete.gaps <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-c">C<a class="anchor" aria-label="anchor" href="#arg-c"></a></dt>
<dd><p>A list of list representing the forest structure. See <code><a href="V.star.html">V.star()</a></code> for more information.</p></dd>


<dt id="arg-zl">ZL<a class="anchor" aria-label="anchor" href="#arg-zl"></a></dt>
<dd><p>A list of integer vectors representing the upper bounds \(\zeta_k\) of the forest structure. See <code><a href="V.star.html">V.star()</a></code> for more information.</p></dd>


<dt id="arg-leaf-list">leaf_list<a class="anchor" aria-label="anchor" href="#arg-leaf-list"></a></dt>
<dd><p>A list of vectors representing the atoms of the forest structure. See <code><a href="V.star.html">V.star()</a></code> for more information.</p></dd>


<dt id="arg-prune-leafs">prune.leafs<a class="anchor" aria-label="anchor" href="#arg-prune-leafs"></a></dt>
<dd><p>A boolean, <code>FALSE</code> by default. If <code>TRUE</code>, will also prune atoms/leafs for which \(\zeta_k \geq |R_k|\),
this makes the computation of <code><a href="V.star.html">V.star()</a></code> and <code><a href="curve.V.star.forest.html">curve.V.star.forest.naive()</a></code> even faster but should not be used with <code><a href="curve.V.star.forest.html">curve.V.star.forest.fast()</a></code>
because this needs the structure to be complete (i.e., with all its atoms). This is why the default option is <code>FALSE</code>.</p></dd>


<dt id="arg-delete-gaps">delete.gaps<a class="anchor" aria-label="anchor" href="#arg-delete-gaps"></a></dt>
<dd><p>A boolean, <code>FALSE</code> by default. If <code>TRUE</code>, will also delete the gaps in the structure induced
by the pruning, see <code><a href="delete.gaps.html">delete.gaps()</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with three named elements.</p><dl><dt><code>VstarNm</code></dt>
<dd><p>\(V^*(\mathbb N_m)\) is computed as by-product by the algorithm, so we might as well return it.</p></dd>

<dt><code>C</code></dt>
<dd><p>The new <code>C</code> after pruning.</p></dd>

<dt><code>ZL</code></dt>
<dd><p>The new <code>ZL</code> after pruning.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Durand, G., Blanchard, G., Neuvial, P., &amp; Roquain, E. (2020). Post hoc false positive control for structured hypotheses. Scandinavian Journal of Statistics, 47(4), 1114-1148.</p>
<p>Durand G. (2025). A fast algorithm to compute a curve of confidence upper bounds for the False Discovery Proportion using a reference family with a forest structure. arXiv:2502.03849.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

