<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sanssouci">
<title>Confidence curves for structured hypotheses • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Confidence curves for structured hypotheses">
<meta property="og:description" content="sanssouci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.13.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a>
    <a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a>
    <a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a>
    <a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a>
    <a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a>
    <a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a>
    <a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a>
    <a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Confidence curves for structured hypotheses</h1>
                        <h4 data-toc-skip class="author">G. Durand, G.
Blanchard, P. Neuvial, E. Roquain</h4>
            
            <h4 data-toc-skip class="date">2024-06-24</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/HEAD/vignettes/CopyOfconfidenceCurves_localized.Rmd" class="external-link"><code>vignettes/CopyOfconfidenceCurves_localized.Rmd</code></a></small>
      <div class="d-none name"><code>CopyOfconfidenceCurves_localized.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The goal of this vignette is to illustrate the post hoc bounds on the
number of true/false positives proposed in <span class="citation">Durand
et al. (2020)</span> for localized signals. More specifically, we
reproduce one of the plots of Figure 12 in <span class="citation">Durand
et al. (2020)</span> using the R package <code>sanssouci</code>. We will
explicitly quote <span class="citation">Durand et al. (2020)</span> when
relevant.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="objective">Objective<a class="anchor" aria-label="anchor" href="#objective"></a>
</h2>
<p>We consider <span class="math inline">\(m\)</span> null ordered
hypotheses partitioned in intervals of size <span class="math inline">\(s\)</span>. For simplicity we set the number of
intervals to be a power of 2: <span class="math inline">\(m = s
2^q\)</span> for some integer <span class="math inline">\(q\)</span>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">7</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">s</span><span class="op">*</span><span class="fl">2</span><span class="op">^</span><span class="va">q</span></span></code></pre></div>
<p>Therefore, we have <span class="math inline">\(m = 1.28\times
10^{4}\)</span>. Our goal is to compare three post hoc bounds. These
bounds are obtained by interpolation from a <em>reference family</em>
where the amount of signal is estimated by probabilistic inequalities,
following the general principle laid down by <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>, and they
differ by the choice of the reference family:</p>
<ul>
<li><p>“Simes”: the bound derived from the <span class="citation">Simes
(1986)</span> inequality as proposed by <span class="citation">Goeman
and Solari (2011)</span> and further studied by <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>. This
bound was introduced in a context where the signal is not
localized.</p></li>
<li><p>“Tree” and “Partition”: two bounds derived from the DKWM
inequality (<span class="citation">Dvoretzky, Kiefer, and Wolfowitz
(1956)</span>, <span class="citation">Massart (1990)</span>), as
proposed in <span class="citation">Durand et al. (2020)</span>. For the
“Partition” bound, the reference family is the original partition <span class="math inline">\((P_k)_k\)</span> of the <span class="math inline">\(m\)</span> null hypotheses into <span class="math inline">\(K=2^q\)</span> intervals. For the “Tree” bound,
the reference family is the perfect binary tree whose leaves are the
elements of the original partition.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="settings">Settings<a class="anchor" aria-label="anchor" href="#settings"></a>
</h2>
<p>We define the following numerical parameters, which characterize the
true/false null hypothesis configuration considered in Section 5 of
<span class="citation">Durand et al. (2020)</span>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K1</span> <span class="op">&lt;-</span> <span class="fl">8</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fl">0.9</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span>  <span class="va">r</span><span class="op">*</span><span class="va">K1</span><span class="op">*</span><span class="va">s</span></span>
<span><span class="va">barmu</span> <span class="op">&lt;-</span> <span class="fl">3</span></span></code></pre></div>
<p>More precisely, quoting <span class="citation">Durand et al.
(2020)</span>:</p>
<blockquote>
<p>the false null hypotheses are contained in <span class="math inline">\(P_k\)</span> for <span class="math inline">\(1
\leq k \leq K_1\)</span>, for some fixed value of <span class="math inline">\(K_1\)</span>. The quantity <span class="math inline">\(r\)</span> is defined similarly as in (20), as the
fraction of false null hypotheses in those <span class="math inline">\(P_k\)</span>, and is set to <span class="math inline">\(r =0.9\)</span>. All of the other partition pieces
only contain true null hypotheses.</p>
</blockquote>
<p>We start by creating a binary tree structure and generating the
signal:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">family</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dyadic.html">dyadic.from.window.size</a></span><span class="op">(</span><span class="va">m</span>, s <span class="op">=</span> <span class="va">s</span>, method <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen.mu.leaves.html">gen.mu.leaves</a></span><span class="op">(</span>m <span class="op">=</span> <span class="va">m</span>, K1 <span class="op">=</span> <span class="va">K1</span>, d <span class="op">=</span> <span class="va">r</span>, grouped <span class="op">=</span> <span class="cn">TRUE</span>, setting <span class="op">=</span> <span class="st">"const"</span>, </span>
<span>                    barmu <span class="op">=</span> <span class="va">barmu</span>, leaf_list <span class="op">=</span> <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span><span class="op">)</span></span></code></pre></div>
<p>This construction is illustrated in the figure below (Figure 11 in
<span class="citation">Durand et al. (2020)</span>) in the case where
<span class="math inline">\(q=3\)</span>.</p>
<div class="figure">
<img src="DBNR_graph-simu.png" alt="Figure 11 in Durand et al." width="100%"><p class="caption">
Figure 11 in Durand et al.
</p>
</div>
<p>We generate <span class="math inline">\(p\)</span>-values according
to the simulation setup in <span class="citation">Durand et al.
(2020)</span>:</p>
<blockquote>
<p>The true null <span class="math inline">\(p\)</span>-values are
distributed as i.i.d. <span class="math inline">\(\mathcal{N}(0,1)\)</span>, and false null <span class="math inline">\(p\)</span>-values are distributed as i.i.d. <span class="math inline">\(\mathcal{N}(\bar{\mu}, 1)\)</span>, where [<span class="math inline">\(\bar{\mu}= 3\)</span>].</p>
</blockquote>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvalues</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen.p.values.html">gen.p.values</a></span><span class="op">(</span>m <span class="op">=</span> <span class="va">m</span>, mu <span class="op">=</span> <span class="va">mu</span>, rho <span class="op">=</span> <span class="fl">0</span>, alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-confidence-curves">Calculate confidence curves<a class="anchor" aria-label="anchor" href="#calculate-confidence-curves"></a>
</h2>
<p>The confidence level for post hoc inference is set to <span class="math inline">\(\alpha = 0.05\)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span></code></pre></div>
<blockquote>
<p>Below, we will be considering confidence curves of the form <span class="math inline">\((k, V(S_k))_{1 \leq k \leq m}\)</span>, where
<span class="math inline">\(S_k\)</span> is the set of the <span class="math inline">\(k\)</span> smallest <span class="math inline">\(p\)</span>-values (regardless of the ordering
given by the partition). Note that focusing on such sets is a priori
favorable to the Simes bound, for which the elements of the reference
family are among the <span class="math inline">\(S_k\)</span>.</p>
</blockquote>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">pvalues</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="va">m</span></span>
<span><span class="va">stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FP"</span>, <span class="va">m</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="true-number-of-false-positives">1- True number of false positives<a class="anchor" aria-label="anchor" href="#true-number-of-false-positives"></a>
</h3>
<p>The true number of false positives will be called “Oracle” bound in
the plots below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">oracle_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">mu</span><span class="op">[</span><span class="va">ord</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Oracle"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">oracle_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simes-based-confidence-curve">2- Simes-based confidence curve<a class="anchor" aria-label="anchor" href="#simes-based-confidence-curve"></a>
</h3>
<p>Here we use the <span class="citation">Simes (1986)</span> inequality
to bound the number of false positives in each node of the tree, as
proposed by <span class="citation">Goeman and Solari (2011)</span> and
further studied by <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>, both in a context where the signal is not
localized.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simes_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curveMaxFP.html">curveMaxFP</a></span><span class="op">(</span><span class="va">pvalues</span>, <span class="va">alpha</span> <span class="op">*</span> <span class="fl">1</span><span class="op">:</span><span class="va">m</span> <span class="op">/</span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Simes"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Simes"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Simes"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">simes_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dkwm-based-confidence-curve">3- DKWM-based confidence curve<a class="anchor" aria-label="anchor" href="#dkwm-based-confidence-curve"></a>
</h3>
<p>Here we use the DKWM inequality (<span class="citation">Dvoretzky,
Kiefer, and Wolfowitz (1956)</span>, <span class="citation">Massart
(1990)</span>) to bound the number of false positives in each node of
the tree, as suggested in <span class="citation">Durand et al.
(2020)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DKWM(tree)"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Tree"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">C_tree</span> <span class="op">&lt;-</span> <span class="va">family</span><span class="op">$</span><span class="va">C</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="va">ZL_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zetas.tree.html">zetas.tree</a></span><span class="op">(</span><span class="va">C_tree</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, <span class="va">zeta.DKWM</span>, <span class="va">pvalues</span>, <span class="va">alpha</span><span class="op">)</span></span>
<span><span class="va">tree_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curve.V.star.forest.html">curve.V.star.forest.fast</a></span><span class="op">(</span><span class="va">ord</span>, <span class="va">C_tree</span>, <span class="va">ZL_tree</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, pruning <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Tree"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">tree_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DKWM(partition)"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Partition"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">ZL_part</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zetas.tree.html">zetas.tree</a></span><span class="op">(</span><span class="va">family</span><span class="op">$</span><span class="va">C</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, <span class="va">zeta.DKWM</span>, <span class="va">pvalues</span>, <span class="va">alpha</span><span class="op">)</span></span>
<span><span class="va">part_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curve.V.star.forest.html">curve.V.star.forest.fast</a></span><span class="op">(</span><span class="va">ord</span>, <span class="va">family</span><span class="op">$</span><span class="va">C</span>, <span class="va">ZL_part</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, pruning <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Partition"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">part_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="plot-confidence-curves">Plot confidence curves<a class="anchor" aria-label="anchor" href="#plot-confidence-curves"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">res</span><span class="op">)</span></span>
<span><span class="va">lvls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="st">"Partition"</span>, <span class="st">"Simes"</span>, <span class="st">"Tree"</span>, <span class="st">"Hybrid"</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lvls</span><span class="op">)</span>, <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lvls</span></span></code></pre></div>
<div class="section level3">
<h3 id="upper-bound-on-the-number-of-false-positives">Upper bound on the number of false positives<a class="anchor" aria-label="anchor" href="#upper-bound-on-the-number-of-false-positives"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xymax</span> <span class="op">&lt;-</span> <span class="va">m1</span>;</span>
<span><span class="va">pV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bound</span>, colour <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Upper bound on the number of false positives"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"sorted hypotheses"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">pV</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/plot-confidence-curve-1.png" width="700"></p>
<p>The “Tree” and “Partition” bounds are sharper than the “Simes” bound
as soon as we are considering “large” sets of hypotheses. The fact that
the “Tree” and “Partition” bounds are not as sharp as the “Simes” bound
for the first hundred of hypotheses can be explained by our choice of
the ordering of the null hypotheses in the sets <span class="math inline">\(S_k\)</span>, which as discussed above is
favorable to the “Simes” bound.</p>
<p>Zooming on the first 720 null hypotheses (in the order of the <span class="math inline">\(p\)</span>-values):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pV</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">xymax</span><span class="op">)</span>,</span>
<span>                     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">xymax</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/zoom-plot-confidence-curve-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="lower-bound-on-the-number-of-true-positives">Lower bound on the number of true positives<a class="anchor" aria-label="anchor" href="#lower-bound-on-the-number-of-true-positives"></a>
</h2>
<p>The same information can be displayed as a lower bound on the number
of true positives, defined for any <span class="math inline">\(S \subset
\{1 \dots m\}\)</span> by <span class="math inline">\(|S| -
V(S)\)</span>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span><span class="op">$</span><span class="va">S</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">x</span> <span class="op">-</span> <span class="va">dat</span><span class="op">$</span><span class="va">bound</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xmax</span> <span class="op">&lt;-</span> <span class="va">m1</span>;</span>
<span><span class="va">ymax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">S</span><span class="op">[</span><span class="va">m1</span><span class="op">]</span><span class="op">)</span>;</span>
<span><span class="va">pS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">S</span>, colour <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Lower bound on the number of true positives"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"sorted hypotheses"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">pS</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/plot-lower-bound-1.png" width="700"></p>
<p>Zooming in the first 720 null hypotheses (in the order of the <span class="math inline">\(p\)</span>-values), we recover the middle plot in
Figure 12 of <span class="citation">Durand et al. (2020)</span>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pS</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">xmax</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">ymax</span><span class="op">)</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/zoom-plot-lower-bound-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] ggplot2_3.5.1    sanssouci_0.13.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-0       gtable_0.3.5       jsonlite_1.8.8     compiler_4.4.1    </span></span>
<span><span class="co">##  [5] highr_0.11         Rcpp_1.0.12        jquerylib_0.1.4    systemfonts_1.1.0 </span></span>
<span><span class="co">##  [9] scales_1.3.0       textshaping_0.4.0  yaml_2.3.8         fastmap_1.2.0     </span></span>
<span><span class="co">## [13] lattice_0.22-6     R6_2.5.1           labeling_0.4.3     generics_0.1.3    </span></span>
<span><span class="co">## [17] knitr_1.47         tibble_3.2.1       desc_1.4.3         munsell_0.5.1     </span></span>
<span><span class="co">## [21] RColorBrewer_1.1-3 pillar_1.9.0       bslib_0.7.0        rlang_1.1.4       </span></span>
<span><span class="co">## [25] utf8_1.2.4         cachem_1.1.0       xfun_0.45          fs_1.6.4          </span></span>
<span><span class="co">## [29] sass_0.4.9         memoise_2.0.1      cli_3.6.3          withr_3.0.0       </span></span>
<span><span class="co">## [33] pkgdown_2.0.9      magrittr_2.0.3     matrixTests_0.2.3  digest_0.6.35     </span></span>
<span><span class="co">## [37] grid_4.4.1         lifecycle_1.0.4    vctrs_0.6.5        evaluate_0.24.0   </span></span>
<span><span class="co">## [41] glue_1.7.0         farver_2.1.2       ragg_1.3.2         fansi_1.0.6       </span></span>
<span><span class="co">## [45] colorspace_2.1-0   rmarkdown_2.27     purrr_1.0.2        pkgconfig_2.0.3   </span></span>
<span><span class="co">## [49] matrixStats_1.3.0  tools_4.4.1        htmltools_0.5.8.1</span></span></code></pre>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-durand20post-hoc" class="csl-entry">
Durand, Guillermo, Gilles Blanchard, Pierre Neuvial, and Etienne
Roquain. 2020. <span>“<span class="nocase">Post hoc false positive
control for structured hypotheses</span>.”</span> <em>Scandinavian
Journal of Statistics</em>. <a href="https://hal.archives-ouvertes.fr/hal-01829037" class="external-link">https://hal.archives-ouvertes.fr/hal-01829037</a>.
</div>
<div id="ref-dvoretzky1956asymptotic" class="csl-entry">
Dvoretzky, Aryeh, Jack Kiefer, and Jacob Wolfowitz. 1956.
<span>“Asymptotic Minimax Character of the Sample Distribution Function
and of the Classical Multinomial Estimator.”</span> <em>The Annals of
Mathematical Statistics</em> 27 (3): 642–69.
</div>
<div id="ref-GS2011" class="csl-entry">
Goeman, Jelle J, and Aldo Solari. 2011. <span>“Multiple Testing for
Exploratory Research.”</span> <em>Statistical Science</em> 26 (4):
584–97.
</div>
<div id="ref-massart1990tight" class="csl-entry">
Massart, Pascal. 1990. <span>“<span class="nocase">The tight constant in
the Dvoretzky-Kiefer-Wolfowitz inequality</span>.”</span> <em>The Annals
of Probability</em> 18 (3): 1269–83.
</div>
<div id="ref-simes86improved" class="csl-entry">
Simes, R J. 1986. <span>“<span class="nocase">An improved Bonferroni
procedure for multiple tests of significance</span>.”</span>
<em>Biometrika</em> 73 (3): 751–54.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
