<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Confidence curves for structured hypotheses • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Confidence curves for structured hypotheses">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.14.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Confidence curves for structured hypotheses</h1>
                        <h4 data-toc-skip class="author">G. Durand, G.
Blanchard, P. Neuvial, E. Roquain</h4>
            
            <h4 data-toc-skip class="date">2025-05-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/v0.14.0/vignettes/CopyOfconfidenceCurves_localized.Rmd" class="external-link"><code>vignettes/CopyOfconfidenceCurves_localized.Rmd</code></a></small>
      <div class="d-none name"><code>CopyOfconfidenceCurves_localized.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The goal of this vignette is to illustrate the post hoc bounds on the
number of true/false positives proposed in <span class="citation">Durand
et al. (2020)</span> for localized signals. More specifically, we
reproduce one of the plots of Figure 12 in <span class="citation">Durand
et al. (2020)</span> using the R package <code>sanssouci</code>. We will
explicitly quote <span class="citation">Durand et al. (2020)</span> when
relevant.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="objective">Objective<a class="anchor" aria-label="anchor" href="#objective"></a>
</h2>
<p>We consider
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
null ordered hypotheses partitioned in intervals of size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
For simplicity we set the number of intervals to be a power of 2:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>=</mo><mi>s</mi><msup><mn>2</mn><mi>q</mi></msup></mrow><annotation encoding="application/x-tex">m = s 2^q</annotation></semantics></math>
for some integer
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>q</mi><annotation encoding="application/x-tex">q</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">7</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">s</span><span class="op">*</span><span class="fl">2</span><span class="op">^</span><span class="va">q</span></span></code></pre></div>
<p>Therefore, we have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>=</mo><mn>1.28</mn><mo>×</mo><msup><mn>10</mn><mn>4</mn></msup></mrow><annotation encoding="application/x-tex">m = 1.28\times 10^{4}</annotation></semantics></math>.
Our goal is to compare three post hoc bounds. These bounds are obtained
by interpolation from a <em>reference family</em> where the amount of
signal is estimated by probabilistic inequalities, following the general
principle laid down by <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>, and they differ by the choice of the reference
family:</p>
<ul>
<li><p>“Simes”: the bound derived from the <span class="citation">Simes
(1986)</span> inequality as proposed by <span class="citation">Goeman
and Solari (2011)</span> and further studied by <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>. This
bound was introduced in a context where the signal is not
localized.</p></li>
<li><p>“Tree” and “Partition”: two bounds derived from the DKWM
inequality (<span class="citation">Dvoretzky, Kiefer, and Wolfowitz
(1956)</span>, <span class="citation">Massart (1990)</span>), as
proposed in <span class="citation">Durand et al. (2020)</span>. For the
“Partition” bound, the reference family is the original partition
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>P</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>k</mi></msub><annotation encoding="application/x-tex">(P_k)_k</annotation></semantics></math>
of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
null hypotheses into
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mo>=</mo><msup><mn>2</mn><mi>q</mi></msup></mrow><annotation encoding="application/x-tex">K=2^q</annotation></semantics></math>
intervals. For the “Tree” bound, the reference family is the perfect
binary tree whose leaves are the elements of the original
partition.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="settings">Settings<a class="anchor" aria-label="anchor" href="#settings"></a>
</h2>
<p>We define the following numerical parameters, which characterize the
true/false null hypothesis configuration considered in Section 5 of
<span class="citation">Durand et al. (2020)</span>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K1</span> <span class="op">&lt;-</span> <span class="fl">8</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fl">0.9</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span>  <span class="va">r</span><span class="op">*</span><span class="va">K1</span><span class="op">*</span><span class="va">s</span></span>
<span><span class="va">barmu</span> <span class="op">&lt;-</span> <span class="fl">3</span></span></code></pre></div>
<p>More precisely, quoting <span class="citation">Durand et al.
(2020)</span>:</p>
<blockquote>
<p>the false null hypotheses are contained in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mi>k</mi></msub><annotation encoding="application/x-tex">P_k</annotation></semantics></math>
for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>≤</mo><mi>k</mi><mo>≤</mo><msub><mi>K</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">1 \leq k \leq K_1</annotation></semantics></math>,
for some fixed value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mn>1</mn></msub><annotation encoding="application/x-tex">K_1</annotation></semantics></math>.
The quantity
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>r</mi><annotation encoding="application/x-tex">r</annotation></semantics></math>
is defined similarly as in (20), as the fraction of false null
hypotheses in those
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mi>k</mi></msub><annotation encoding="application/x-tex">P_k</annotation></semantics></math>,
and is set to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mo>=</mo><mn>0.9</mn></mrow><annotation encoding="application/x-tex">r =0.9</annotation></semantics></math>.
All of the other partition pieces only contain true null hypotheses.</p>
</blockquote>
<p>We start by creating a binary tree structure and generating the
signal:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">family</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dyadic.html">dyadic.from.window.size</a></span><span class="op">(</span><span class="va">m</span>, s <span class="op">=</span> <span class="va">s</span>, method <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen.mu.leaves.html">gen.mu.leaves</a></span><span class="op">(</span>m <span class="op">=</span> <span class="va">m</span>, K1 <span class="op">=</span> <span class="va">K1</span>, d <span class="op">=</span> <span class="va">r</span>, grouped <span class="op">=</span> <span class="cn">TRUE</span>, setting <span class="op">=</span> <span class="st">"const"</span>, </span>
<span>                    barmu <span class="op">=</span> <span class="va">barmu</span>, leaf_list <span class="op">=</span> <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span><span class="op">)</span></span></code></pre></div>
<p>This construction is illustrated in the figure below (Figure 11 in
<span class="citation">Durand et al. (2020)</span>) in the case where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">q=3</annotation></semantics></math>.</p>
<div class="figure">
<img src="DBNR_graph-simu.png" alt="Figure 11 in Durand et al." width="100%"><p class="caption">
Figure 11 in Durand et al.
</p>
</div>
<p>We generate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
according to the simulation setup in <span class="citation">Durand et
al. (2020)</span>:</p>
<blockquote>
<p>The true null
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
are distributed as i.i.d.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathcal{N}(0,1)</annotation></semantics></math>,
and false null
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
are distributed as i.i.d.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>μ</mi><mo accent="true">‾</mo></mover><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathcal{N}(\bar{\mu}, 1)</annotation></semantics></math>,
where
[<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>μ</mi><mo accent="true">‾</mo></mover><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">\bar{\mu}= 3</annotation></semantics></math>].</p>
</blockquote>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvalues</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen.p.values.html">gen.p.values</a></span><span class="op">(</span>m <span class="op">=</span> <span class="va">m</span>, mu <span class="op">=</span> <span class="va">mu</span>, rho <span class="op">=</span> <span class="fl">0</span>, alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-confidence-curves">Calculate confidence curves<a class="anchor" aria-label="anchor" href="#calculate-confidence-curves"></a>
</h2>
<p>The confidence level for post hoc inference is set to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.05</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span></code></pre></div>
<blockquote>
<p>Below, we will be considering confidence curves of the form
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>k</mi><mo>,</mo><mi>V</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>S</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mo>≤</mo><mi>k</mi><mo>≤</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">(k, V(S_k))_{1 \leq k \leq m}</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mi>k</mi></msub><annotation encoding="application/x-tex">S_k</annotation></semantics></math>
is the set of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
smallest
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
(regardless of the ordering given by the partition). Note that focusing
on such sets is a priori favorable to the Simes bound, for which the
elements of the reference family are among the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mi>k</mi></msub><annotation encoding="application/x-tex">S_k</annotation></semantics></math>.</p>
</blockquote>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">pvalues</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="va">m</span></span>
<span><span class="va">stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FP"</span>, <span class="va">m</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="true-number-of-false-positives">1- True number of false positives<a class="anchor" aria-label="anchor" href="#true-number-of-false-positives"></a>
</h3>
<p>The true number of false positives will be called “Oracle” bound in
the plots below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">oracle_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">mu</span><span class="op">[</span><span class="va">ord</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Oracle"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">oracle_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simes-based-confidence-curve">2- Simes-based confidence curve<a class="anchor" aria-label="anchor" href="#simes-based-confidence-curve"></a>
</h3>
<p>Here we use the <span class="citation">Simes (1986)</span> inequality
to bound the number of false positives in each node of the tree, as
proposed by <span class="citation">Goeman and Solari (2011)</span> and
further studied by <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>, both in a context where the signal is not
localized.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simes_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curveMaxFP.html">curveMaxFP</a></span><span class="op">(</span><span class="va">pvalues</span>, <span class="va">alpha</span> <span class="op">*</span> <span class="fl">1</span><span class="op">:</span><span class="va">m</span> <span class="op">/</span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Simes"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Simes"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Simes"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">simes_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dkwm-based-confidence-curve">3- DKWM-based confidence curve<a class="anchor" aria-label="anchor" href="#dkwm-based-confidence-curve"></a>
</h3>
<p>Here we use the DKWM inequality (<span class="citation">Dvoretzky,
Kiefer, and Wolfowitz (1956)</span>, <span class="citation">Massart
(1990)</span>) to bound the number of false positives in each node of
the tree, as suggested in <span class="citation">Durand et al.
(2020)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DKWM(tree)"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Tree"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">C_tree</span> <span class="op">&lt;-</span> <span class="va">family</span><span class="op">$</span><span class="va">C</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="va">ZL_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zetas.tree.html">zetas.tree</a></span><span class="op">(</span><span class="va">C_tree</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, <span class="va">zeta.DKWM</span>, <span class="va">pvalues</span>, <span class="va">alpha</span><span class="op">)</span></span>
<span><span class="va">tree_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curve.V.star.forest.html">curve.V.star.forest.fast</a></span><span class="op">(</span><span class="va">ord</span>, <span class="va">C_tree</span>, <span class="va">ZL_tree</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, pruning <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Tree"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">tree_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DKWM(partition)"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Partition"</span>, <span class="va">m</span><span class="op">)</span></span>
<span><span class="va">ZL_part</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zetas.tree.html">zetas.tree</a></span><span class="op">(</span><span class="va">family</span><span class="op">$</span><span class="va">C</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, <span class="va">zeta.DKWM</span>, <span class="va">pvalues</span>, <span class="va">alpha</span><span class="op">)</span></span>
<span><span class="va">part_bound</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/curve.V.star.forest.html">curve.V.star.forest.fast</a></span><span class="op">(</span><span class="va">ord</span>, <span class="va">family</span><span class="op">$</span><span class="va">C</span>, <span class="va">ZL_part</span>, <span class="va">family</span><span class="op">$</span><span class="va">leaf_list</span>, pruning <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="st">"Partition"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, label <span class="op">=</span> <span class="va">label</span>, stat <span class="op">=</span> <span class="va">stat</span>, bound <span class="op">=</span> <span class="va">part_bound</span>, method <span class="op">=</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="plot-confidence-curves">Plot confidence curves<a class="anchor" aria-label="anchor" href="#plot-confidence-curves"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">res</span><span class="op">)</span></span>
<span><span class="va">lvls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Oracle"</span>, <span class="st">"Partition"</span>, <span class="st">"Simes"</span>, <span class="st">"Tree"</span>, <span class="st">"Hybrid"</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lvls</span><span class="op">)</span>, <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lvls</span></span></code></pre></div>
<div class="section level3">
<h3 id="upper-bound-on-the-number-of-false-positives">Upper bound on the number of false positives<a class="anchor" aria-label="anchor" href="#upper-bound-on-the-number-of-false-positives"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xymax</span> <span class="op">&lt;-</span> <span class="va">m1</span>;</span>
<span><span class="va">pV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bound</span>, colour <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Upper bound on the number of false positives"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"sorted hypotheses"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">pV</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/plot-confidence-curve-1.png" width="700"></p>
<p>The “Tree” and “Partition” bounds are sharper than the “Simes” bound
as soon as we are considering “large” sets of hypotheses. The fact that
the “Tree” and “Partition” bounds are not as sharp as the “Simes” bound
for the first hundred of hypotheses can be explained by our choice of
the ordering of the null hypotheses in the sets
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mi>k</mi></msub><annotation encoding="application/x-tex">S_k</annotation></semantics></math>,
which as discussed above is favorable to the “Simes” bound.</p>
<p>Zooming on the first 720 null hypotheses (in the order of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pV</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">xymax</span><span class="op">)</span>,</span>
<span>                     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">xymax</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/zoom-plot-confidence-curve-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="lower-bound-on-the-number-of-true-positives">Lower bound on the number of true positives<a class="anchor" aria-label="anchor" href="#lower-bound-on-the-number-of-true-positives"></a>
</h2>
<p>The same information can be displayed as a lower bound on the number
of true positives, defined for any
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mo>⊂</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mi>…</mi><mi>m</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">S \subset \{1 \dots m\}</annotation></semantics></math>
by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">|</mo><mi>S</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo>−</mo><mi>V</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>S</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">|S| - V(S)</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span><span class="op">$</span><span class="va">S</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">$</span><span class="va">x</span> <span class="op">-</span> <span class="va">dat</span><span class="op">$</span><span class="va">bound</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xmax</span> <span class="op">&lt;-</span> <span class="va">m1</span>;</span>
<span><span class="va">ymax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">S</span><span class="op">[</span><span class="va">m1</span><span class="op">]</span><span class="op">)</span>;</span>
<span><span class="va">pS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">S</span>, colour <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Lower bound on the number of true positives"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"sorted hypotheses"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">pS</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/plot-lower-bound-1.png" width="700"></p>
<p>Zooming in the first 720 null hypotheses (in the order of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values),
we recover the middle plot in Figure 12 of <span class="citation">Durand
et al. (2020)</span>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pS</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">xmax</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">ymax</span><span class="op">)</span></span></code></pre></div>
<p><img src="CopyOfconfidenceCurves_localized_files/figure-html/zoom-plot-lower-bound-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] ggplot2_3.5.2    sanssouci_0.14.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-3       gtable_0.3.6       jsonlite_2.0.0     compiler_4.5.0    </span></span>
<span><span class="co">##  [5] Rcpp_1.0.14        jquerylib_0.1.4    systemfonts_1.2.3  scales_1.4.0      </span></span>
<span><span class="co">##  [9] textshaping_1.0.1  yaml_2.3.10        fastmap_1.2.0      lattice_0.22-6    </span></span>
<span><span class="co">## [13] R6_2.6.1           labeling_0.4.3     generics_0.1.3     knitr_1.50        </span></span>
<span><span class="co">## [17] tibble_3.2.1       desc_1.4.3         bslib_0.9.0        pillar_1.10.2     </span></span>
<span><span class="co">## [21] RColorBrewer_1.1-3 rlang_1.1.6        cachem_1.1.0       xfun_0.52         </span></span>
<span><span class="co">## [25] fs_1.6.6           sass_0.4.10        cli_3.6.5          pkgdown_2.1.2     </span></span>
<span><span class="co">## [29] withr_3.0.2        magrittr_2.0.3     matrixTests_0.2.3  digest_0.6.37     </span></span>
<span><span class="co">## [33] grid_4.5.0         lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.3    </span></span>
<span><span class="co">## [37] glue_1.8.0         farver_2.1.2       ragg_1.4.0         rmarkdown_2.29    </span></span>
<span><span class="co">## [41] matrixStats_1.5.0  tools_4.5.0        pkgconfig_2.0.3    htmltools_0.5.8.1</span></span></code></pre>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-durand20post-hoc" class="csl-entry">
Durand, Guillermo, Gilles Blanchard, Pierre Neuvial, and Etienne
Roquain. 2020. <span>“<span class="nocase">Post hoc false positive
control for structured hypotheses</span>.”</span> <em>Scandinavian
Journal of Statistics</em>. <a href="https://hal.archives-ouvertes.fr/hal-01829037" class="external-link">https://hal.archives-ouvertes.fr/hal-01829037</a>.
</div>
<div id="ref-dvoretzky1956asymptotic" class="csl-entry">
Dvoretzky, Aryeh, Jack Kiefer, and Jacob Wolfowitz. 1956.
<span>“Asymptotic Minimax Character of the Sample Distribution Function
and of the Classical Multinomial Estimator.”</span> <em>The Annals of
Mathematical Statistics</em> 27 (3): 642–69.
</div>
<div id="ref-GS2011" class="csl-entry">
Goeman, Jelle J, and Aldo Solari. 2011. <span>“Multiple Testing for
Exploratory Research.”</span> <em>Statistical Science</em> 26 (4):
584–97.
</div>
<div id="ref-massart1990tight" class="csl-entry">
Massart, Pascal. 1990. <span>“<span class="nocase">The tight constant in
the Dvoretzky-Kiefer-Wolfowitz inequality</span>.”</span> <em>The Annals
of Probability</em> 18 (3): 1269–83.
</div>
<div id="ref-simes86improved" class="csl-entry">
Simes, R J. 1986. <span>“<span class="nocase">An improved Bonferroni
procedure for multiple tests of significance</span>.”</span>
<em>Biometrika</em> 73 (3): 751–54.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
