<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>IIDEA: Interactive Inference for Differential Expression Analyses • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="IIDEA: Interactive Inference for Differential Expression Analyses">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.14.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>IIDEA: Interactive Inference for Differential Expression Analyses</h1>
                        <h4 data-toc-skip class="author">Nicolas
Enjalbert-Courrech and Pierre Neuvial</h4>
            
            <h4 data-toc-skip class="date">2025-05-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/v0.14.0/vignettes/IIDEA.Rmd" class="external-link"><code>vignettes/IIDEA.Rmd</code></a></small>
      <div class="d-none name"><code>IIDEA.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides guidelines for using the <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr" class="external-link">IIDEA shiny
application</a>. In the context of differential gene expression studies,
this interactive application combines volcano plots with enrichment
analyses.</p>
<p>The main distinctive feature of this app is that it makes it possible
for users to interactively select genes of interest (either from a
volcano plot or corresponding to a particular gene set) and obtain valid
statistical guarantees regarding the number of true/false positives
among these genes (see Methods section below).</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<ul>
<li><p>Open the <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">IIDEA shiny
application</a></p></li>
<li><p>Click on ‘Run!’</p></li>
<li><p>This launches the analysis of an example data set (described
below): a volcano plot is displayed in the right panel, together with a
table at the bottom of the left panel.</p></li>
</ul>
<p><img src="img/IIDEA_overview.png" width="600px"></p>
</div>
<div class="section level2">
<h2 id="interactive-statistical-guarantees">Interactive statistical guarantees<a class="anchor" aria-label="anchor" href="#interactive-statistical-guarantees"></a>
</h2>
<div class="section level3">
<h3 id="default-output">Default output<a class="anchor" aria-label="anchor" href="#default-output"></a>
</h3>
<p><img src="img/IIDEA_post-hoc-table.png" width="400px"></p>
<p>The table displays, for each user selection of a set of genes, the
following statistical guarantees:</p>
<ul>
<li>
<strong>TP<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>≥</mo><annotation encoding="application/x-tex">\geq</annotation></semantics></math></strong>:
there are at least 57 truly differentially expressed genes (True
Positives, TP) among the 70 genes selected in the volcano plot.</li>
<li>
<strong>FDP<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>≤</mo><annotation encoding="application/x-tex">\leq</annotation></semantics></math></strong>:
the proportion of genes erroneously called differentially expressed
among those selected (False Discovery Proportion, FDP) is less than
19%.</li>
</ul>
</div>
<div class="section level3">
<h3 id="user-selections">User selections<a class="anchor" aria-label="anchor" href="#user-selections"></a>
</h3>
<ul>
<li>Select a set of genes of interest by dragging and dropping the
horizonal and vertical bars, or by making a rectangular or a “lasso”
selection. The table is updated with this selection.</li>
<li>Clicking on the line corresponding to a gene selection highlights
the corresponding genes in the volcano plot.</li>
<li>Clicking on the name of a gene selection in the table opens a new
window with the corresponding set of genes displayed by <a href="https://string-db.org/" class="external-link">StringDB</a>.</li>
<li>User selections and post hoc bounds can be downloaded as CSV
files.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="enrichment-analyses">Enrichment analyses<a class="anchor" aria-label="anchor" href="#enrichment-analyses"></a>
</h2>
<ul>
<li>Click on the ‘Gene sets’ tab in the left panel: the table at the
bottom of the left panel now displays statistical guarantees obtained by
post hoc bounds for a list of GO terms (corresponding to the Biological
Process ontology in the example below).</li>
<li>Clicking on the line corresponding to GO a term highlights the
corresponding genes in the volcano plot (see next Figure).</li>
<li>Clicking on a GO term opens the associated QuickGO page (<span class="citation">Binns et al. (2009)</span>), e.g. <a href="https://www.ebi.ac.uk/QuickGO/term/GO:0002576" class="external-link uri">https://www.ebi.ac.uk/QuickGO/term/GO:0002576</a>.</li>
<li>Post hoc bounds can be downloaded as a CSV file.</li>
</ul>
<p>By default, all available gene sets are listed. As noted by <span class="citation">Ebrahimpoor et al. (2020)</span>, post hoc inference
makes it possible to perform enrichment analyses for both the
“self-contained” and “competitive” hypotheses. Quoting <span class="citation">Ebrahimpoor et al. (2020)</span>:</p>
<ul>
<li>“Methods testing the competitive null test whether features in the
feature set of interest are more often active than features outside the
feature set.</li>
<li>Methods testing the self-contained null test whether there are any
active features in the feature set of interest.”</li>
</ul>
<p>Both approaches are readily avialable via a drop-down menu, which
makes it possible to filter only those gene sets that are significant
according to the “self-contained” or “competitive” null hypotheses.</p>
<p>For example, in the Leukemia data set, the GO term Platelet
degranulation (<a href="GO:0002576" class="uri">GO:0002576</a>) is
significant both for the competitive and the self-contained null
hypotheses. This GO term is highlighted in the next plot, which has been
obtained by clicking on the associated line in the gene set table.</p>
<p><img src="img/IIDEA_GO-0002576.png" width="600px"></p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<div class="section level3">
<h3 id="public-data">Public data<a class="anchor" aria-label="anchor" href="#public-data"></a>
</h3>
<div class="section level4">
<h4 id="leukemia-data-set">Leukemia data set<a class="anchor" aria-label="anchor" href="#leukemia-data-set"></a>
</h4>
<p>The first data set available in the list of public data sets is the
Leukemia data set studied in <span class="citation">Bourgon, Gentleman,
and Huber (2010)</span>.</p>
<p>This data set consists of gene expression measurements for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>=</mo><mn>79</mn></mrow><annotation encoding="application/x-tex">n = 79</annotation></semantics></math>
patients with B-cell acute lymphoblastic leukemia (ALL) <span class="citation">Chiaretti et al. (2005)</span>. These patients are
classified into two subgoups, depending on whether or not they harbor a
specific mutation called “BCR/ABL”. A Welch
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>test
of equality of group mean expression is applied to each gene.</p>
<p>This data set is available from the R package <a href="https://github.com/sanssouci-org/sanssouci.data" class="external-link"><code>sanssouci.data</code></a>,
and also studied in the <a href="post-hoc_differential-expression.html">differential expression
vignette</a> of the SansSouci package.</p>
</div>
<div class="section level4">
<h4 id="gseabenchmarker-data">GSEAbenchmarkeR data<a class="anchor" aria-label="anchor" href="#gseabenchmarker-data"></a>
</h4>
<p>Nine other data sets from <a href="https://bioconductor.org/packages/release/bioc/html/GSEABenchmarkeR.html" class="external-link"><code>GSEABenchmarkeR</code></a>
package are also available. These data sets contain gene expression
measurements from microarray experiments for various diseases (several
cancer types, Alzheimer’s disease). Each of these data sets can be
choosen for analysis via ‘Choose a gene data set’ dropdown menu.</p>
<p>Thanks to the Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/EnrichmentBrowser.html" class="external-link"><code>EnrichmentBrowser</code></a>,
an enrichement analysis of gene sets based on Gene Ontology gene sets
(Biological Process) is available.</p>
</div>
</div>
<div class="section level3">
<h3 id="user-data">User data<a class="anchor" aria-label="anchor" href="#user-data"></a>
</h3>
<p>If “Use public data” is unchecked, users can upload their own gene
expression data, and optionally a file containing gene sets of interest
(for example obtained from Gene Ontology).</p>
<div class="section level4">
<h4 id="gene-expression">Gene expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h4>
<p>Gene expression data must be in of the two following formats (see
also the example data files that can be downloaded from IIDEA):</p>
<ul>
<li><p>A matrix containing gene expression data, in the form of a gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>×</mo><annotation encoding="application/x-tex">\times</annotation></semantics></math>
condition matrix. The first row of this matrix should be a header
containing category labels 0 or 1. The first column contains an unique
gene identifier per row. One current limitation of IIDEA in this case is
that it performs Welch
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
tests for differential expression, which is only suitable for microarray
data;</p></li>
<li><p>A matrix containing
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
and fold changes for each gene. This matrix can be obtained from any
standard differential expression method (e.g. DESeq2, edgeR,
limma-voom).</p></li>
</ul>
<p>Gene expression data must be stored in a csv file. the first row must
contain the name of the columns. The separator should be a ‘,’.
Nevertheless, ‘;’ and ’ are allowed.</p>
<p>From a statistical perspective, an important limitation when only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
and fold changes are provided is that post hoc inference can only be
performed using the original post hoc bound of <span class="citation">Goeman and Solari (2011)</span>, and not using the more
powerful bounds of <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>. Indeed, the latter require the full expression
matrix to be available in order to use permutations and obtain bounds
that are adaptive to the dependence in the data set at hand. An
important perspective for IIDEA is to directly implement differential
expression methods dedicated to count data, so that the powerful bounds
of <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>
can be used.</p>
</div>
<div class="section level4">
<h4 id="gene-sets">Gene sets<a class="anchor" aria-label="anchor" href="#gene-sets"></a>
</h4>
<p>Gene set information must be a CSV file where each row is a gene and
each column is a gene set. It is a binary matrix specifying which gene
is contained in each gene set. The first row must be a header containing
the gene set names, and it must contain the same gene identifiers as the
gene expression matrix.</p>
<p>While only the gene expression matrix is mandatory to obtain
interactive volcano plots, uploading the gene set matrix is required for
gene set analyses. Sample input data files can be downloaded as an
example of the expected file structure. These files correspond to the
public data set <span class="citation">Chiaretti et al.
(2005)</span>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-the-app-on-a-local-computer">Running the app on a local computer<a class="anchor" aria-label="anchor" href="#running-the-app-on-a-local-computer"></a>
</h2>
<p>IIDEA is deployed at . No login is required, and the data sets
uploaded by users are not stored on the server side. The source code of
the R package which contains the source files for IIDEA, is available at
. This allows users to run IIDEA on a local machine or deploy it on
their own servers if desired. Once the ‘remotes’ package is installed,
IIDEA can be run locally from an R session by executing the following
commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"sanssouci-org/sanssouci"</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu">runApp</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-examples/volcano-plot"</span>, package <span class="op">=</span> <span class="st">"sanssouci"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>By default, only the Leukemia data set is available for a local
installation. To add the nine other public data sets, you may run the
following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-examples/volcano-plot/GSEABenchmarkeR/export-data.R"</span>, </span>
<span>                   package <span class="op">=</span> <span class="st">"sanssouci"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This code uses functions of the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABenchmarkeR.html" class="external-link">GSEABenchmarkeR</a>
package to normalize and store locally gene expression data sets and
corresponding gene annotation from GO Biological Process. Please note
that it may take several minutes to run (but it only needs to be run
once).</p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>IIDEA leverages post hoc inference methods implemented in the R
package <a href="https://github.com/sanssouci-org/sanssouci" class="external-link"><code>sanssouci</code></a>.</p>
<p>Post hoc inference is a recent field stemming from the multiple
testing literature where users are allowed to freely select hypotheses
of interest after “seeing the data”, without compromising statistical
validity despite the issue of “cherry-picking” (<span class="citation">Blanchard, Neuvial, and Roquain (2021)</span>, <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>, <span class="citation">Goeman and Solari (2011)</span>).</p>
<p>The methods implemented by IIDEA allow for valid inference without
requiring assumptions on the dependence between genes. Instead, such
dependencies are <em>estimated</em> from the data using permutation
techniques, making our methods both mathematically valid and
statistically powerful (see e.g. <a href="https://sanssouci-org.github.io/sanssouci/articles/post-hoc_differential-expression.html" class="uri">https://sanssouci-org.github.io/sanssouci/articles/post-hoc_differential-expression.html</a>).</p>
<p>The parameters for this estimation step can be customized by clicking
“Advanced parameters”.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-binns2009quickgo" class="csl-entry">
Binns, David, Emily Dimmer, Rachael Huntley, Daniel Barrell, Claire
O’donovan, and Rolf Apweiler. 2009. <span>“QuickGO: A Web-Based Tool for
Gene Ontology Searching.”</span> <em>Bioinformatics</em> 25 (22):
3045–46.
</div>
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-BNR:chap" class="csl-entry">
———. 2021. <span>“<span class="nocase">On agnostic post hoc approaches
to false positive control</span>.”</span> In <em><span class="nocase">Handbook of Multiple Comparisons</span></em>, edited by
Xinping Cui, Thorsten Dickhaus, Ying Ding, and Jason C. Hsu. Handbooks
of Modern Statistical Methods. <span>Chapman<br>
&amp; Hall/CRC</span>. <a href="https://hal.archives-ouvertes.fr/hal-02320543" class="external-link">https://hal.archives-ouvertes.fr/hal-02320543</a>.
</div>
<div id="ref-BGH2010" class="csl-entry">
Bourgon, Richard, Robert Gentleman, and Wolfgang Huber. 2010.
<span>“<span class="nocase">Independent filtering increases detection
power for high-throughput experiments</span>.”</span> <em>PNAS</em>. <a href="https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107" class="external-link">https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107</a>.
</div>
<div id="ref-CLGV+2005" class="csl-entry">
Chiaretti, Sabina, Xiaochun Li, Robert Gentleman, Antonella Vitale,
Kathy S Wang, Franco Mandelli, Robin Foa, and Jerome Ritz. 2005.
<span>“Gene Expression Profiles of b-Lineage Adult Acute Lymphocytic
Leukemia Reveal Genetic Patterns That Identify Lineage Derivation and
Distinct Mechanisms of Transformation.”</span> <em>Clinical Cancer
Research</em> 11 (20): 7209–19.
</div>
<div id="ref-ebrahimpoor2020simultaneous" class="csl-entry">
Ebrahimpoor, Mitra, Pietro Spitali, Kristina Hettne, Roula Tsonaka, and
Jelle Goeman. 2020. <span>“Simultaneous Enrichment Analysis of All
Possible Gene-Sets: Unifying Self-Contained and Competitive
Methods.”</span> <em>Briefings in Bioinformatics</em> 21 (4): 1302–12.
</div>
<div id="ref-GS2011" class="csl-entry">
Goeman, Jelle J, and Aldo Solari. 2011. <span>“Multiple Testing for
Exploratory Research.”</span> <em>Statistical Science</em> 26 (4):
584–97.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
