<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>IIDEA: Interactive Inference for Differential Expression Analyses • sanssouci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="IIDEA: Interactive Inference for Differential Expression Analyses">
<meta property="og:description" content="sanssouci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sanssouci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a>
    </li>
    <li>
      <a href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a>
    </li>
    <li>
      <a href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a>
    </li>
    <li>
      <a href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a>
    </li>
    <li>
      <a href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a>
    </li>
    <li>
      <a href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a>
    </li>
    <li>
      <a href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/sanssouci/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="IIDEA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>IIDEA: Interactive Inference for Differential Expression Analyses</h1>
                        <h4 data-toc-skip class="author">Nicolas Enjalbert-Courrech and Pierre Neuvial</h4>
            
            <h4 data-toc-skip class="date">2022-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pneuvial/sanssouci/blob/HEAD/vignettes/IIDEA.Rmd" class="external-link"><code>vignettes/IIDEA.Rmd</code></a></small>
      <div class="hidden name"><code>IIDEA.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides guidelines for using the <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr" class="external-link">IIDEA shiny application</a>. In the context of differential gene expression studies, this interactive application combines volcano plots with enrichment analyses.</p>
<p>The main distinctive feature of this app is that it makes it possible for users to interactively select genes of interest (either from a volcano plot or corresponding to a particular gene set) and obtain valid statistical guarantees regarding the number of true/false positives among these genes (see Methods section below).</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<ul>
<li><p>Open the <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">IIDEA shiny application</a></p></li>
<li><p>Click on ‘Run!’</p></li>
<li><p>This launches the analysis of an example data set (described below): a volcano plot is displayed in the right panel, together with a table at the bottom of the left panel.</p></li>
</ul>
<p><img src="img/IIDEA_overview.png" width="600px"></p>
</div>
<div class="section level2">
<h2 id="interactive-statistical-guarantees">Interactive statistical guarantees<a class="anchor" aria-label="anchor" href="#interactive-statistical-guarantees"></a>
</h2>
<div class="section level3">
<h3 id="default-output">Default output<a class="anchor" aria-label="anchor" href="#default-output"></a>
</h3>
<p><img src="img/IIDEA_post-hoc-table.png" width="400px"></p>
<p>The table displays, for each user selection of a set of genes, the following statistical guarantees:</p>
<ul>
<li>
<strong>TP<span class="math inline">\(\geq\)</span></strong>: there are at least 57 truly differentially expressed genes (True Positives, TP) among the 70 genes selected in the volcano plot.</li>
<li>
<strong>FDP<span class="math inline">\(\leq\)</span></strong>: the proportion of genes erroneously called differentially expressed among those selected (False Discovery Proportion, FDP) is less than 19%.</li>
</ul>
</div>
<div class="section level3">
<h3 id="user-selections">User selections<a class="anchor" aria-label="anchor" href="#user-selections"></a>
</h3>
<ul>
<li>Select a set of genes of interest by dragging and dropping the horizonal and vertical bars, or by making a rectangular or a “lasso” selection. The table is updated with this selection.</li>
<li>Clicking on the line corresponding to a gene selection highlights the corresponding genes in the volcano plot.</li>
<li>Clicking on the name of a gene selection in the table opens a new window with the corresponding set of genes displayed by <a href="https://string-db.org/" class="external-link">StringDB</a>.</li>
<li>User selections and post hoc bounds can be downloaded as CSV files.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="enrichment-analyses">Enrichment analyses<a class="anchor" aria-label="anchor" href="#enrichment-analyses"></a>
</h2>
<ul>
<li>Click on the ‘Gene sets’ tab in the left panel: the table at the bottom of the left panel now displays statistical guarantees obtained by post hoc bounds for a list of GO terms (corresponding to the Biological Process ontology in the example below).</li>
<li>Clicking on the line corresponding to GO a term highlights the corresponding genes in the volcano plot (see next Figure).</li>
<li>Clicking on a GO term opens the associated QuickGO page (<span class="citation">Binns et al. (2009)</span>), e.g. <a href="https://www.ebi.ac.uk/QuickGO/term/GO:0002576" class="external-link uri">https://www.ebi.ac.uk/QuickGO/term/GO:0002576</a>.</li>
<li>Post hoc bounds can be downloaded as a CSV file.</li>
</ul>
<p>By default, all available gene sets are listed. As noted by <span class="citation">Ebrahimpoor et al. (2020)</span>, post hoc inference makes it possible to perform enrichment analyses for both the “self-contained” and “competitive” hypotheses. Quoting <span class="citation">Ebrahimpoor et al. (2020)</span>:</p>
<ul>
<li>"Methods testing the competitive null test whether features in the feature set of interest are more often active than features outside the feature set.</li>
<li>Methods testing the self-contained null test whether there are any active features in the feature set of interest."</li>
</ul>
<p>Both approaches are readily avialable via a drop-down menu, which makes it possible to filter only those gene sets that are significant according to the “self-contained” or “competitive” null hypotheses.</p>
<p>For example, in the Leukemia data set, the GO term Platelet degranulation (<a href="GO:0002576" class="uri">GO:0002576</a>) is significant both for the competitive and the self-contained null hypotheses. This GO term is highlighted in the next plot, which has been obtained by clicking on the associated line in the gene set table.</p>
<p><img src="img/IIDEA_GO-0002576.png" width="600px"></p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<div class="section level3">
<h3 id="public-data">Public data<a class="anchor" aria-label="anchor" href="#public-data"></a>
</h3>
<div class="section level4">
<h4 id="leukemia-data-set">Leukemia data set<a class="anchor" aria-label="anchor" href="#leukemia-data-set"></a>
</h4>
<p>The first data set available in the list of public data sets is the Leukemia data set studied in <span class="citation">Bourgon, Gentleman, and Huber (2010)</span>.</p>
<p>This data set consists of gene expression measurements for <span class="math inline">\(n = 79\)</span> patients with B-cell acute lymphoblastic leukemia (ALL) <span class="citation">Chiaretti et al. (2005)</span>. These patients are classified into two subgoups, depending on whether or not they harbor a specific mutation called “BCR/ABL”. A Welch <span class="math inline">\(T\)</span>test of equality of group mean expression is applied to each gene.</p>
<p>This data set is available from the R package <a href="https://github.com/pneuvial/sanssouci.data" class="external-link"><code>sanssouci.data</code></a>, and also studied in the <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression.html">differential expression vignette</a> of the SansSouci package.</p>
</div>
<div class="section level4">
<h4 id="gseabenchmarker-data">GSEAbenchmarkeR data<a class="anchor" aria-label="anchor" href="#gseabenchmarker-data"></a>
</h4>
<p>Nine other data sets from <a href="https://bioconductor.org/packages/release/bioc/html/GSEABenchmarkeR.html" class="external-link"><code>GSEABenchmarkeR</code></a> package are also available. These data sets contain gene expression measurements from microarray experiments for various diseases (several cancer types, Alzheimer’s disease). Each of these data sets can be choosen for analysis via ‘Choose a gene data set’ dropdown menu.</p>
<p>Thanks to the Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/EnrichmentBrowser.html" class="external-link"><code>EnrichmentBrowser</code></a>, an enrichement analysis of gene sets based on Gene Ontology gene sets (Biological Process) is available.</p>
</div>
</div>
<div class="section level3">
<h3 id="user-data">User data<a class="anchor" aria-label="anchor" href="#user-data"></a>
</h3>
<p>If “Use public data” is unchecked, users can upload their own gene expression data, and optionally a file containing gene sets of interest (for example obtained from Gene Ontology).</p>
<div class="section level4">
<h4 id="gene-expression">Gene expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h4>
<p>Gene expression data must be in of the two following formats (see also the example data files that can be downloaded from IIDEA):</p>
<ul>
<li><p>A matrix containing gene expression data, in the form of a gene <span class="math inline">\(\times\)</span> condition matrix. The first row of this matrix should be a header containing category labels 0 or 1. The first column contains an unique gene identifier per row. One current limitation of IIDEA in this case is that it performs Welch <span class="math inline">\(T\)</span> tests for differential expression, which is only suitable for microarray data;</p></li>
<li><p>A matrix containing <span class="math inline">\(p\)</span>-values and fold changes for each gene. This matrix can be obtained from any standard differential expression method (e.g. DESeq2, edgeR, limma-voom).</p></li>
</ul>
<p>Gene expression data must be stored in a csv file. the first row must contain the name of the columns. The separator should be a ‘,’. Nevertheless, ‘;’ and ’ are allowed.</p>
<p>From a statistical perspective, an important limitation when only <span class="math inline">\(p\)</span>-values and fold changes are provided is that post hoc inference can only be performed using the original post hoc bound of <span class="citation">Goeman and Solari (2011)</span>, and not using the more powerful bounds of <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>. Indeed, the latter require the full expression matrix to be available in order to use permutations and obtain bounds that are adaptive to the dependence in the data set at hand. An important perspective for IIDEA is to directly implement differential expression methods dedicated to count data, so that the powerful bounds of <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span> can be used.</p>
</div>
<div class="section level4">
<h4 id="gene-sets">Gene sets<a class="anchor" aria-label="anchor" href="#gene-sets"></a>
</h4>
<p>Gene set information must be a CSV file where each row is a gene and each column is a gene set. It is a binary matrix specifying which gene is contained in each gene set. The first row must be a header containing the gene set names, and it must contain the same gene identifiers as the gene expression matrix.</p>
<p>While only the gene expression matrix is mandatory to obtain interactive volcano plots, uploading the gene set matrix is required for gene set analyses. Sample input data files can be downloaded as an example of the expected file structure. These files correspond to the public data set <span class="citation">Chiaretti et al. (2005)</span>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-the-app-on-a-local-computer">Running the app on a local computer<a class="anchor" aria-label="anchor" href="#running-the-app-on-a-local-computer"></a>
</h2>
<p>IIDEA is deployed at . No login is required, and the data sets uploaded by users are not stored on the server side. The source code of the R package  which contains the source files for IIDEA, is available at . This allows users to run IIDEA on a local machine or deploy it on their own servers if desired. Once the ‘remotes’ package is installed, IIDEA can be run locally from an R session by executing the following commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pneuvial/sanssouci"</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu">runApp</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-examples/volcano-plot"</span>, package <span class="op">=</span> <span class="st">"sanssouci"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>By default, only the Leukemia data set is available for a local installation. To add the nine other public data sets, you may run the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"shiny-examples/volcano-plot/GSEABenchmarkeR/export-data.R"</span>, </span>
<span>                   package <span class="op">=</span> <span class="st">"sanssouci"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This code uses functions of the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABenchmarkeR.html" class="external-link">GSEABenchmarkeR</a> package to normalize and store locally gene expression data sets and corresponding gene annotation from GO Biological Process. Please note that it may take several minutes to run (but it only needs to be run once).</p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>IIDEA leverages post hoc inference methods implemented in the R package <a href="https://github.com/pneuvial/sanssouci" class="external-link"><code>sanssouci</code></a>.</p>
<p>Post hoc inference is a recent field stemming from the multiple testing literature where users are allowed to freely select hypotheses of interest after “seeing the data”, without compromising statistical validity despite the issue of “cherry-picking” (<span class="citation">Blanchard, Neuvial, and Roquain (2021)</span>, <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>, <span class="citation">Goeman and Solari (2011)</span>).</p>
<p>The methods implemented by IIDEA allow for valid inference without requiring assumptions on the dependence between genes. Instead, such dependencies are <em>estimated</em> from the data using permutation techniques, making our methods both mathematically valid and statistically powerful (see e.g. <a href="https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression.html" class="uri">https://pneuvial.github.io/sanssouci/articles/post-hoc_differential-expression.html</a>).</p>
<p>The parameters for this estimation step can be customized by clicking “Advanced parameters”.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-binns2009quickgo">
<p>Binns, David, Emily Dimmer, Rachael Huntley, Daniel Barrell, Claire O’donovan, and Rolf Apweiler. 2009. “QuickGO: A Web-Based Tool for Gene Ontology Searching.” <em>Bioinformatics</em> 25 (22): 3045–6.</p>
</div>
<div id="ref-blanchard20post-hoc">
<p>Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020. “Post Hoc Confidence Bounds on False Positives Using Reference Families.” <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.</p>
</div>
<div id="ref-BNR:chap">
<p>———. 2021. “On agnostic post hoc approaches to false positive control.” In <em>Handbook of Multiple Comparisons</em>, edited by Xinping Cui, Thorsten Dickhaus, Ying Ding, and Jason C. Hsu. Handbooks of Modern Statistical Methods. Chapman<br>
&amp; Hall/CRC. <a href="https://hal.archives-ouvertes.fr/hal-02320543" class="external-link">https://hal.archives-ouvertes.fr/hal-02320543</a>.</p>
</div>
<div id="ref-BGH2010">
<p>Bourgon, Richard, Robert Gentleman, and Wolfgang Huber. 2010. “Independent filtering increases detection power for high-throughput experiments.” <em>PNAS</em>. <a href="https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107" class="external-link">https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107</a>.</p>
</div>
<div id="ref-CLGV+2005">
<p>Chiaretti, Sabina, Xiaochun Li, Robert Gentleman, Antonella Vitale, Kathy S Wang, Franco Mandelli, Robin Foa, and Jerome Ritz. 2005. “Gene Expression Profiles of B-Lineage Adult Acute Lymphocytic Leukemia Reveal Genetic Patterns That Identify Lineage Derivation and Distinct Mechanisms of Transformation.” <em>Clinical Cancer Research</em> 11 (20): 7209–19.</p>
</div>
<div id="ref-ebrahimpoor2020simultaneous">
<p>Ebrahimpoor, Mitra, Pietro Spitali, Kristina Hettne, Roula Tsonaka, and Jelle Goeman. 2020. “Simultaneous Enrichment Analysis of All Possible Gene-Sets: Unifying Self-Contained and Competitive Methods.” <em>Briefings in Bioinformatics</em> 21 (4): 1302–12.</p>
</div>
<div id="ref-GS2011">
<p>Goeman, Jelle J, and Aldo Solari. 2011. “Multiple Testing for Exploratory Research.” <em>Statistical Science</em> 26 (4): 584–97.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.math.univ-toulouse.fr/~pneuvial/" class="external-link">Pierre Neuvial</a>, <a href="https://www.imo.universite-paris-saclay.fr/~blanchard/" class="external-link">Gilles Blanchard</a>, <a href="https://durandg12.github.io/" class="external-link">Guillermo Durand</a>, Nicolas Enjalbert-Courrech, <a href="https://etienneroquain-81.webself.net/" class="external-link">Etienne Roquain</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
