<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Joint Error Rate calibration ‚Ä¢ sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Joint Error Rate calibration">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.14.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Joint Error Rate calibration</h1>
            <h3 data-toc-skip class="subtitle">Simulations for one and
two-sample tests</h3>
                        <h4 data-toc-skip class="author">P. Neuvial</h4>
            
            <h4 data-toc-skip class="date">2018-03-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/v0.14.0/vignettes/jointErrorRateCalibration_simulations.Rmd" class="external-link"><code>vignettes/jointErrorRateCalibration_simulations.Rmd</code></a></small>
      <div class="d-none name"><code>jointErrorRateCalibration_simulations.Rmd</code></div>
    </div>

    
    
<p>This vignettes illustrates the following points</p>
<ul>
<li>simulation of one- and two-sample Gaussian equi-correlated
observations</li>
<li>computation of test statistics by randomization</li>
<li>calibration of Joint-Family-Wise Error Rate (JER) thresholds</li>
</ul>
<p>We start with two-sample tests because we believe they are used more
frequently.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span>
<span><span class="co">#set.seed(0xBEEF) # for reproducibility</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span></code></pre></div>
<p>Parameters:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">5e2</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">pi0</span> <span class="op">&lt;-</span> <span class="fl">0.8</span></span>
<span><span class="va">rho</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span><span class="va">SNR</span> <span class="op">&lt;-</span> <span class="fl">3</span></span></code></pre></div>
<p>We use the function <code>SansSouciSim</code> to generate Gaussian
equi-correlated samples. We then ‚Äòfit‚Äô the resulting SansSouci
object</p>
<div class="section level2">
<h2 id="two-sample-tests">Two-sample tests<a class="anchor" aria-label="anchor" href="#two-sample-tests"></a>
</h2>
<div class="section level3">
<h3 id="simulation">Simulation<a class="anchor" aria-label="anchor" href="#simulation"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouciSim.html">SansSouciSim</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">rho</span>, <span class="va">n</span>, <span class="va">pi0</span>, SNR <span class="op">=</span> <span class="va">SNR</span>, prob <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">obj</span></span></code></pre></div>
<pre><code><span><span class="co">## 'SansSouci' object:</span></span>
<span><span class="co">##  Number of hypotheses:    500 </span></span>
<span><span class="co">##  Number of observations:  30 </span></span>
<span><span class="co">##  2-sample data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Truth: </span></span>
<span><span class="co">##   100 false null hypotheses (signals) out of 500 (pi0=0.8)</span></span></code></pre>
<p>We perform JER calibration using the linear (Simes) template
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mi>‚Ñú</mi><mrow><mo stretchy="true" form="prefix" mathvariant="fraktur">(</mo><mi mathvariant="fraktur">Œª</mi><mo stretchy="true" form="postfix" mathvariant="fraktur">)</mo></mrow></mrow><mo>=</mo><mo stretchy="false" form="prefix">(</mo><msub><mi>R</mi><mn>1</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œª</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">\mathfrak{R(\lambda)}=(R_1(\lambda),</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>‚Ä¶</mi><mo>,</mo><msub><mi>R</mi><mi>K</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œª</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\ldots,R_K(\lambda)</annotation></semantics></math>,
where for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>‚â§</mo><mi>k</mi><mo>‚â§</mo><mi>K</mi></mrow><annotation encoding="application/x-tex">1 \leq k \leq K</annotation></semantics></math></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>R</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œª</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mrow><mo stretchy="true" form="prefix">{</mo><mi>i</mi><mo>‚àà</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>‚Ä¶</mi><mo>,</mo><mi>m</mi><mo stretchy="false" form="postfix">}</mo><mspace width="0.222em"></mspace><mo>:</mo><mspace width="0.222em"></mspace><mover><mi>Œ¶</mi><mo accent="true">‚Äæ</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Z</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>&gt;</mo><mfrac><mrow><mi>Œª</mi><mi>k</mi></mrow><mi>m</mi></mfrac><mo stretchy="true" form="postfix">}</mo></mrow><mspace width="0.167em"></mspace></mrow><annotation encoding="application/x-tex">R_k(\lambda) = \left\{i\in \{1, \dots , m\}\::\: \bar{\Phi}(Z_i) &gt;  \frac{\lambda k}{m} \right\}\,</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>Œ¶</mi><mo accent="true">‚Äæ</mo></mover><annotation encoding="application/x-tex">\bar{\Phi}</annotation></semantics></math>
is the cdf of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùí©</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathcal{N}(0,1)</annotation></semantics></math>
distribution. Note that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>i</mi></msub><mo>=</mo><mover><mi>Œ¶</mi><mo accent="true">‚Äæ</mo></mover><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Z</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p_i = \bar{\Phi}(Z_i)</annotation></semantics></math>
is the one-sided
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
associated to the test statistics
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Z</mi><mi>i</mi></msub><annotation encoding="application/x-tex">Z_i</annotation></semantics></math>.</p>
</div>
<div class="section level3">
<h3 id="calibration">Calibration<a class="anchor" aria-label="anchor" href="#calibration"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.2</span></span>
<span><span class="va">cal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="va">B</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span>
<span><span class="va">cal</span></span></code></pre></div>
<pre><code><span><span class="co">## 'SansSouci' object:</span></span>
<span><span class="co">##  Number of hypotheses:    500 </span></span>
<span><span class="co">##  Number of observations:  30 </span></span>
<span><span class="co">##  2-sample data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Truth: </span></span>
<span><span class="co">##   100 false null hypotheses (signals) out of 500 (pi0=0.8)</span></span>
<span><span class="co">## Parameters: </span></span>
<span><span class="co">##  Test function:      rowWelchTests</span></span>
<span><span class="co">##  Number of permutations: B=1000</span></span>
<span><span class="co">##  Significance level: alpha=0.2</span></span>
<span><span class="co">##  Reference family:   Simes</span></span>
<span><span class="co">##      (of size:   K=500)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Output:</span></span>
<span><span class="co">##  Calibration parameter:  lambda=0.2051593</span></span></code></pre>
<p>The output of the calibration is the following <code>SansSouci</code>
object. In particular:</p>
<ul>
<li>
<code>cal$input</code> contains the input simulated data (and the
associated truth)</li>
<li>
<code>cal$param</code> contains the calibration parameters</li>
<li>
<code>cal$output</code> contains the output of the calibration,
including
<ul>
<li>
<code>p.values</code>:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
test statistics calculated by permutation</li>
<li>
<code>thr</code> : A JER-controlling family of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
(here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mo>=</mo><mi>m</mi></mrow><annotation encoding="application/x-tex">K=m</annotation></semantics></math>)
elements</li>
<li>
<code>lambda</code>: the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œª</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>-calibration
parameter</li>
</ul>
</li>
</ul>
<p>Because we are under positive equi-correlation, we expect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œª</mi><mo>&gt;</mo><mi>Œ±</mi></mrow><annotation encoding="application/x-tex">\lambda &gt; \alpha</annotation></semantics></math>
for the Simes family.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cal</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">lambda</span></span></code></pre></div>
<pre><code><span><span class="co">##       20% </span></span>
<span><span class="co">## 0.2051593</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cal</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">lambda</span> <span class="op">&gt;</span> <span class="va">alpha</span></span></code></pre></div>
<pre><code><span><span class="co">##  20% </span></span>
<span><span class="co">## TRUE</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="post-hoc-confidence-bounds">Post hoc confidence bounds<a class="anchor" aria-label="anchor" href="#post-hoc-confidence-bounds"></a>
</h3>
<p>The fitted <code>SansSouci</code> object contains post hoc confidence
bounds:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cal</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">env</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   x label stat bound</span></span>
<span><span class="co">## 1 1 Simes   TP     1</span></span>
<span><span class="co">## 2 2 Simes   TP     2</span></span>
<span><span class="co">## 3 3 Simes   TP     3</span></span>
<span><span class="co">## 4 4 Simes   TP     4</span></span>
<span><span class="co">## 5 5 Simes   TP     5</span></span>
<span><span class="co">## 6 6 Simes   TP     6</span></span></code></pre>
<p>We compare it to the true number of false positives among the most
significant items, and to the Simes bound without calibration</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cal0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span>
<span><span class="va">oracle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, family <span class="op">=</span> <span class="st">"Oracle"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">confs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Simes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cal0</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>              <span class="st">"Simes+calibration"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cal</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>              <span class="st">"Oracle"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">oracle</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">confs</span>, xmax <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="jointErrorRateCalibration_simulations_files/figure-html/plot-conf-env-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="one-sample-tests">One sample tests<a class="anchor" aria-label="anchor" href="#one-sample-tests"></a>
</h2>
<p>The code is identical, except for the line to generate the
observations (where we do not specify a probability of belonging to one
of the two populations using the <code>prob</code> argument); moreover
it is not necessary to specify a vector of categories ‚Äòcateg‚Äô in
‚ÄòcalibrateJER‚Äô.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouciSim.html">SansSouciSim</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">rho</span>, <span class="va">n</span>, <span class="va">pi0</span>, SNR <span class="op">=</span> <span class="va">SNR</span><span class="op">)</span></span>
<span><span class="va">obj</span></span></code></pre></div>
<pre><code><span><span class="co">## 'SansSouci' object:</span></span>
<span><span class="co">##  Number of hypotheses:    500 </span></span>
<span><span class="co">##  Number of observations:  30 </span></span>
<span><span class="co">##  1-sample data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Truth: </span></span>
<span><span class="co">##   100 false null hypotheses (signals) out of 500 (pi0=0.8)</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="va">B</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span>
<span><span class="va">cal</span></span></code></pre></div>
<pre><code><span><span class="co">## 'SansSouci' object:</span></span>
<span><span class="co">##  Number of hypotheses:    500 </span></span>
<span><span class="co">##  Number of observations:  30 </span></span>
<span><span class="co">##  1-sample data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Truth: </span></span>
<span><span class="co">##   100 false null hypotheses (signals) out of 500 (pi0=0.8)</span></span>
<span><span class="co">## Parameters: </span></span>
<span><span class="co">##  Test function:      rowZTests</span></span>
<span><span class="co">##  Number of permutations: B=1000</span></span>
<span><span class="co">##  Significance level: alpha=0.2</span></span>
<span><span class="co">##  Reference family:   Simes</span></span>
<span><span class="co">##      (of size:   K=500)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Output:</span></span>
<span><span class="co">##  Calibration parameter:  lambda=0.382704</span></span></code></pre>
<p>Again we expect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œª</mi><mo>&gt;</mo><mi>Œ±</mi></mrow><annotation encoding="application/x-tex">\lambda &gt; \alpha</annotation></semantics></math>.</p>
<div class="section level3">
<h3 id="confidence-curves">Confidence curves<a class="anchor" aria-label="anchor" href="#confidence-curves"></a>
</h3>
<p>The associated confidence curves are displayed below:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cal0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span>
<span><span class="va">oracle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, family <span class="op">=</span> <span class="st">"Oracle"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">confs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Simes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cal0</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>              <span class="st">"Simes+calibration"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cal</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>              <span class="st">"Oracle"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">oracle</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">confs</span><span class="op">)</span></span></code></pre></div>
<p><img src="jointErrorRateCalibration_simulations_files/figure-html/plot-conf-env-one-sample-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] ggplot2_3.5.2    sanssouci_0.14.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-3       gtable_0.3.6       jsonlite_2.0.0     compiler_4.5.0    </span></span>
<span><span class="co">##  [5] Rcpp_1.0.14        jquerylib_0.1.4    systemfonts_1.2.3  scales_1.4.0      </span></span>
<span><span class="co">##  [9] textshaping_1.0.1  yaml_2.3.10        fastmap_1.2.0      lattice_0.22-6    </span></span>
<span><span class="co">## [13] R6_2.6.1           labeling_0.4.3     generics_0.1.3     knitr_1.50        </span></span>
<span><span class="co">## [17] tibble_3.2.1       desc_1.4.3         bslib_0.9.0        pillar_1.10.2     </span></span>
<span><span class="co">## [21] RColorBrewer_1.1-3 rlang_1.1.6        cachem_1.1.0       xfun_0.52         </span></span>
<span><span class="co">## [25] fs_1.6.6           sass_0.4.10        cli_3.6.5          pkgdown_2.1.2     </span></span>
<span><span class="co">## [29] withr_3.0.2        magrittr_2.0.3     matrixTests_0.2.3  digest_0.6.37     </span></span>
<span><span class="co">## [33] grid_4.5.0         lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.3    </span></span>
<span><span class="co">## [37] glue_1.8.0         farver_2.1.2       ragg_1.4.0         rmarkdown_2.29    </span></span>
<span><span class="co">## [41] matrixStats_1.5.0  tools_4.5.0        pkgconfig_2.0.3    htmltools_0.5.8.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
