<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Permutation-based post hoc inference for fMRI studies • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Permutation-based post hoc inference for fMRI studies">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.16.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_linear-model.html">Bootstrap-based post hoc inference for differential gene expression studies in linear models</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for two-group differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for two-group differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Permutation-based post hoc inference for fMRI studies</h1>
                        <h4 data-toc-skip class="author">Alexandre Blain
and Pierre Neuvial</h4>
            
            <h4 data-toc-skip class="date">2025-07-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/master/vignettes/post-hoc_fMRI.Rmd" class="external-link"><code>vignettes/post-hoc_fMRI.Rmd</code></a></small>
      <div class="d-none name"><code>post-hoc_fMRI.Rmd</code></div>
    </div>

    
    
<p>This vignette shows how to use the <a href="https://github.com/sanssouci-org/sanssouci" class="external-link"><code>sanssouci</code></a>
package to perform permutation-based post hoc inference for fMRI data.
We focus on a motor task (left versus right click) from the Localizer
data set by <span class="citation">Orfanos et al. (2017)</span>. The
data set is included in the <a href="https://github.com/sanssouci-org/sanssouci.data" class="external-link"><code>sanssouci.data</code></a>
package. It has been obtained using the Python module <a href="http://nilearn.github.io/" class="external-link">nilearn</a>, see <span class="citation">Abraham et al. (2014)</span>.</p>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>We first make sure that the required R packages are installed and
loaded.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"sanssouci-org/sanssouci@develop"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/github-org/sanssouci.data" class="external-link">"sanssouci.data"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"sanssouci-org/sanssouci.data"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span></code></pre></div>
<p>We set the seed of the random number generator for numerical
reproducibility of the results:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">20200924</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-and-preprocessing-the-data">Loading and preprocessing the data<a class="anchor" aria-label="anchor" href="#loading-and-preprocessing-the-data"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">fMRI_localizer</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">fMRI_localizer</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">fMRI_localizer</span><span class="op">)</span></span></code></pre></div>
<p>The samples are classified into two subgoups, These subgroups
correspond to two different motor tasks the subjects were asked to
perform, “left click” vs “right click”:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="va">categ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">categ</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">categ</th>
<th align="right">Freq</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">left</td>
<td align="right">15</td>
</tr>
<tr class="even">
<td align="left">right</td>
<td align="right">15</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="post-hoc-inference-by-the-simes-inequality">Post hoc inference by the Simes inequality<a class="anchor" aria-label="anchor" href="#post-hoc-inference-by-the-simes-inequality"></a>
</h2>
<p>We start by calculating the Simes bound introduced by <span class="citation">Goeman and Solari (2011)</span>. This bound has already
been applied to fMRI data, see <span class="citation">Rosenblatt et al.
(2018)</span>. In that paper, post hoc inference (by the Simes
inequality) is referred to as “All-resolutions inference” (ARI).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span></code></pre></div>
<p>We set the target confidence level to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.1</annotation></semantics></math>.
The Simes bound is implemented in the package ‘cherry’. Here, we use the
‘sanssouci’ package, as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span> <span class="op">==</span> <span class="st">"left"</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># map to 0/1</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">Y</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span></span>
<span><span class="va">res_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span>, alpha <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span> <span class="co">## B=0 =&gt; no calibration!</span></span></code></pre></div>
<p>Post hoc bounds are obtained from the <code>bound</code>
function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TP_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res_Simes</span>, what <span class="op">=</span> <span class="st">"TP"</span><span class="op">)</span></span>
<span><span class="va">TP_Simes</span></span>
<span><span class="co">#&gt; [1] 88</span></span></code></pre></div>
<p>The Simes bound implies that with probability larger than 0.9, the
number of voxels associated to the task of interest (among all 4.89^{4}
voxels) is at least 88.</p>
<p>Disregarding (for now) the voxel locations in the brain, we plot
confidence bounds on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
level sets (corresponding to maps obtained by thresholding at a given
test statistic):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xmax</span> <span class="op">&lt;-</span> <span class="fl">800</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res_Simes</span>, xmax <span class="op">=</span> <span class="va">xmax</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Post hoc confidence curves: Simes/ARI"</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_fMRI_files/figure-html/plot-conf-bounds-Simes-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="tighter-confidence-bounds-by-adaptation-to-unknown-dependence">Tighter confidence bounds by adaptation to unknown dependence<a class="anchor" aria-label="anchor" href="#tighter-confidence-bounds-by-adaptation-to-unknown-dependence"></a>
</h2>
<p>As discussed in <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>, the above-described bound is known to be valid
only under certain positive dependence assumptions (PRDS) on the joint
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
distribution. Although the PRDS assumption is widely accepted for fMRI
studies (see <span class="citation">Genovese, Lazar, and Nichols
(2002)</span>, <span class="citation">Nichols and Hayasaka
(2003)</span>), we argue (and demonstrate below) that this assumption
yields overly conservative post hoc bounds. Indeed, the Simes bound is
by construction not <em>adaptive</em> to the specific type of dependence
at hand for a particular data set.</p>
<p>To bypass these limitations, <span class="citation">Blanchard,
Neuvial, and Roquain (2020)</span> have proposed a randomization-based
procedure known as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>-calibration,
which yields tighter bounds that are adapted to the dependency observed
in the data set at hand. We note that a related approach has been
proposed by <span class="citation">Hemerik, Solari, and Goeman
(2019)</span>. In the case of two-sample tests, this calibration can be
achieved by permutation of class labels, which is readily available via
the <code>fit</code> function of the sanssouci package:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="va">B</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span></code></pre></div>
<p>An alternative to the Simes/Linear reference family is the Beta
reference family:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">res_Beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">res</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="va">B</span>, family <span class="op">=</span> <span class="st">"Beta"</span>, K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span></code></pre></div>
<p>As expected from the theory, the post hoc bounds obtained after
calibration by these methods is much tighter than the Simes bound:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Simes/ARI"</span> <span class="op">=</span> <span class="va">res_Simes</span>,</span>
<span>            <span class="st">"Linear"</span> <span class="op">=</span> <span class="va">res</span>,</span>
<span>            <span class="st">"Beta"</span> <span class="op">=</span> <span class="va">res_Beta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">resList</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Beta (K=%s)"</span>, <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, what <span class="op">=</span> <span class="st">"TP"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">bounds</span>, digits <span class="op">=</span> <span class="fl">2</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lower bound on True Positives"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Lower bound on True Positives</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Simes/ARI</td>
<td align="right">88</td>
</tr>
<tr class="even">
<td align="left">Linear</td>
<td align="right">321</td>
</tr>
<tr class="odd">
<td align="left">Beta (K=500)</td>
<td align="right">480</td>
</tr>
</tbody>
</table>
<p>In the next two sections we illustrate the use of these improved
bounds in order to build</p>
<ul>
<li>confidence curves for the true or false positives</li>
<li>confidence statements for brain atlases</li>
</ul>
</div>
<div class="section level2">
<h2 id="confidence-curves-on-top-k-lists">Confidence curves on
“top-<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>”
lists<a class="anchor" aria-label="anchor" href="#confidence-curves-on-top-k-lists"></a>
</h2>
<p>In the absence of prior information on voxels, a natural idea is to
rank them by decreasing statistical significance, and a natural question
to ask is:</p>
<blockquote>
<p>Can we provide a lower confidence curve on the number (or proportion)
of truly associated voxels among the most significant ones?</p>
</blockquote>
<p>We illustrate the use of post-hoc methods to provide this type of
information. More specifcally, we build confidence statements on the
number of true/false positives within the top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
most significant voxels, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
may be defined by the user after seing the data, and multiple choices of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
are allowed.</p>
<p>The confidence curves obtained by calibration of the Simes and Beta
families can be compared graphically to the (parametric) Simes curve
that can be obtained from <span class="citation">Goeman and Solari
(2011)</span>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conf_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># cols &lt;- c("lightgray", "black", "darkgray")</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conf_bounds</span><span class="op">)</span>, <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">conf_bounds</span>, xmax <span class="op">=</span> <span class="fl">700</span>, cols <span class="op">=</span> <span class="va">cols</span>, legend.title <span class="op">=</span> <span class="st">"Template"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="post-hoc_fMRI_files/figure-html/plot-conf-bounds-1.png" width="768"></p>
<p>In this example, ARI is outperformed by permutation-based approaches,
which are able to adapt to unknown dependency: for the same target risk
level
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.1</annotation></semantics></math>),
both permutation-based bounds are much less conservative than the
classical Simes/ARI bound.</p>
</div>
<div class="section level2">
<h2 id="post-hoc-bounds-on-brain-atlas-areas">Post hoc bounds on brain atlas areas<a class="anchor" aria-label="anchor" href="#post-hoc-bounds-on-brain-atlas-areas"></a>
</h2>
<p>The goal in this section is to calculate post hoc bound on
user-defined brain regions. One definition of such regions is given by
the <a href="https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/Atlases" class="external-link">Harvard-Oxford
brain atlas</a>, which is available from the sanssouci.data package.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">brainAtlas</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span></code></pre></div>
<p>This atlas is made of 48 areas. For each of these areas, we calculate
the three post hoc bounds obtained above. Note that by construction of
post hoc bounds, these 48 bounds are valid simultaneously (in
particular, no further multiple testing correction is required at this
stage).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">brainAtlas</span>, <span class="fl">1</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">reg</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">reg</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Region size"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">S</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, <span class="va">S</span>, <span class="st">"TP"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>In this particular case, some signal is revealed by the post hoc
bound for 3 areas for permutation-based approaches:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ww</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">bounds</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">cap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Lower bounds on the number of true positives in %s brain regions."</span>,</span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ww</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">bounds</span><span class="op">[</span>, <span class="va">ww</span><span class="op">]</span><span class="op">)</span>, caption <span class="op">=</span> <span class="va">cap</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Lower bounds on the number of true positives in 3 brain
regions.</caption>
<colgroup>
<col width="48%">
<col width="14%">
<col width="12%">
<col width="8%">
<col width="16%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">Region size</th>
<th align="right">Simes/ARI</th>
<th align="right">Linear</th>
<th align="right">Beta (K=500)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Precentral Gyrus</td>
<td align="right">6640</td>
<td align="right">79</td>
<td align="right">242</td>
<td align="right">298</td>
</tr>
<tr class="even">
<td align="left">Postcentral Gyrus</td>
<td align="right">5030</td>
<td align="right">69</td>
<td align="right">194</td>
<td align="right">209</td>
</tr>
<tr class="odd">
<td align="left">Supramarginal Gyrus, anterior division</td>
<td align="right">2017</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>Here again, ARI is outperformed by permutation-based approaches which
are able to adapt to unknown dependency: for the same target risk level
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.1</annotation></semantics></math>)
both permutation-based bounds are substantially less conservative than
the classical Simes/ARI bound.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] sanssouci_0.16.0 ggplot2_3.5.2   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.7-3       gtable_0.3.6       jsonlite_2.0.0     compiler_4.5.1    </span></span>
<span><span class="co">#&gt;  [5] crayon_1.5.3       jquerylib_0.1.4    systemfonts_1.2.3  scales_1.4.0      </span></span>
<span><span class="co">#&gt;  [9] textshaping_1.0.1  yaml_2.3.10        fastmap_1.2.0      lattice_0.22-7    </span></span>
<span><span class="co">#&gt; [13] R6_2.6.1           labeling_0.4.3     generics_0.1.4     knitr_1.50        </span></span>
<span><span class="co">#&gt; [17] tibble_3.3.0       desc_1.4.3         bslib_0.9.0        pillar_1.11.0     </span></span>
<span><span class="co">#&gt; [21] RColorBrewer_1.1-3 rlang_1.1.6        cachem_1.1.0       xfun_0.52         </span></span>
<span><span class="co">#&gt; [25] fs_1.6.6           sass_0.4.10        cli_3.6.5          pkgdown_2.1.3     </span></span>
<span><span class="co">#&gt; [29] withr_3.0.2        magrittr_2.0.3     matrixTests_0.2.3  digest_0.6.37     </span></span>
<span><span class="co">#&gt; [33] grid_4.5.1         lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.4    </span></span>
<span><span class="co">#&gt; [37] glue_1.8.0         farver_2.1.2       codetools_0.2-20   ragg_1.4.0        </span></span>
<span><span class="co">#&gt; [41] rmarkdown_2.29     matrixStats_1.5.0  tools_4.5.1        pkgconfig_2.0.3   </span></span>
<span><span class="co">#&gt; [45] htmltools_0.5.8.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>To re-build this vignette from its source, use:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_fMRI.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"pdf_document"</span><span class="op">)</span></span>
<span><span class="co"># To keep intermediate files, add option 'clean = FALSE'</span></span>
<span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_fMRI.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"html_document"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-abraham2014machine-learning" class="csl-entry">
Abraham, Alexandre, Fabian Pedregosa, Michael Eickenberg, Philippe
Gervais, Andreas Mueller, Jean Kossaifi, Alexandre Gramfort, Bertrand
Thirion, and Gaël Varoquaux. 2014. <span>“Machine Learning for
Neuroimaging with Scikit-Learn.”</span> <em>Frontiers in
Neuroinformatics</em> 8: 14.
</div>
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-genovese2002thresholding" class="csl-entry">
Genovese, Christopher R, Nicole A Lazar, and Thomas Nichols. 2002.
<span>“Thresholding of Statistical Maps in Functional Neuroimaging Using
the False Discovery Rate.”</span> <em>Neuroimage</em> 15 (4): 870–78.
</div>
<div id="ref-GS2011" class="csl-entry">
Goeman, Jelle J, and Aldo Solari. 2011. <span>“Multiple Testing for
Exploratory Research.”</span> <em>Statistical Science</em> 26 (4):
584–97.
</div>
<div id="ref-HSG2019" class="csl-entry">
Hemerik, Jesse, Aldo Solari, and Jelle J Goeman. 2019.
<span>“Permutation-Based Simultaneous Confidence Bounds for the False
Discovery Proportion.”</span> <em>Biometrika</em> 106 (3): 635–49.
</div>
<div id="ref-nichols2003controlling" class="csl-entry">
Nichols, Thomas, and Satoru Hayasaka. 2003. <span>“Controlling the
Familywise Error Rate in Functional Neuroimaging: A Comparative
Review.”</span> <em>Statistical Methods in Medical Research</em> 12 (5):
419–46.
</div>
<div id="ref-orfanos2017brainomics" class="csl-entry">
Orfanos, Dimitri Papadopoulos, Vincent Michel, Yannick Schwartz,
Philippe Pinel, Antonio Moreno, Denis Le Bihan, and Vincent Frouin.
2017. <span>“The Brainomics/Localizer Database.”</span>
<em>Neuroimage</em> 144: 309–14.
</div>
<div id="ref-rosenblatt2018ari" class="csl-entry">
Rosenblatt, Jonathan D, Livio Finos, Wouter D Weeda, Aldo Solari, and
Jelle J Goeman. 2018. <span>“All-Resolutions Inference for Brain
Imaging.”</span> <em>Neuroimage</em> 181: 786–96.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
