<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Bootstrap-based post hoc inference for differential gene expression studies in linear models • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Bootstrap-based post hoc inference for differential gene expression studies in linear models">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.16.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_linear-model.html">Bootstrap-based post hoc inference for differential gene expression studies in linear models</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for two-group differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for two-group differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Bootstrap-based post hoc inference for differential gene expression studies in linear models</h1>
                        <h4 data-toc-skip class="author">Nicolas
Enjalbert Courrech, Pierre Neuvial</h4>
            
            <h4 data-toc-skip class="date">2025-07-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/master/vignettes/post-hoc_differential-expression_linear-model.Rmd" class="external-link"><code>vignettes/post-hoc_differential-expression_linear-model.Rmd</code></a></small>
      <div class="d-none name"><code>post-hoc_differential-expression_linear-model.Rmd</code></div>
    </div>

    
    
<p>This vignette illustrates the relevance of bootstrap-based post hoc
bounds on false positives for the differential analysis of gene
expression measurements (microarray data). We illustrate how the post
hoc methods introduced by <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span> and extended to linear model by <span class="citation">Davenport, Thirion, and Neuvial (2025)</span> may be
used to</p>
<ul>
<li>build confidence curves (envelopes) for the true or false positives
and defined differentially expressed genes accordingly</li>
<li>perform statistical inference on the output of volcano plots</li>
</ul>
<p>The methods described in this vignette have been extended by <span class="citation">Davenport, Thirion, and Neuvial (2025)</span> for
complex experimental design witch need linear models. In this vignette
we reproduce some of the results obtained in Section 5.2 of <span class="citation">Davenport, Thirion, and Neuvial (2025)</span>.
<!-- A shiny application for volcano plots is also available from  https://shiny-iidea-sanssouci.apps.math.cnrs.fr/. --></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"sanssouci-org/sanssouci@bootstrap-calibration"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/github-org/sanssouci.data" class="external-link">"sanssouci.data"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">"matrixStats"</a></span><span class="op">)</span></span></code></pre></div>
<p>Set the seed of the random number generator for numerical
reproducibility of the results:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">20250705</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="motivation-a-differential-gene-expression-study">Motivation: a differential gene expression study<a class="anchor" aria-label="anchor" href="#motivation-a-differential-gene-expression-study"></a>
</h2>
<p>We focus on differential gene expression studies in Chronic
Obstructive Pulmonary Disease (COPD), see <span class="citation">Bahr et
al. (2013)</span>, whose main goal was to identify genes whose
expression level is significantly associated with lung function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"expr_COPD"</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">expr_COPD</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"expr_COPD_design"</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">expr_COPD_design</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">!=</span><span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">keep.nonNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/anyMissing.html" class="external-link">rowAnyMissings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span>, <span class="va">keep.nonNA</span><span class="op">]</span></span>
<span>  <span class="va">X</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[</span><span class="va">keep.nonNA</span>, <span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The data consists of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mo>=</mo><mn>12531</mn></mrow><annotation encoding="application/x-tex">D = 12531</annotation></semantics></math>
gene expression measurements for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>=</mo><mn>135</mn></mrow><annotation encoding="application/x-tex">n = 135</annotation></semantics></math>
patients. For each gene, we fit a linear model including the pulmonary
obstruction ratio (<code>fev1_fvc_utah</code>), along with the following
covariates (see also <code><a href="https://rdrr.io/pkg/sanssouci.data/man/expr_COPD_design.html" class="external-link">?sanssouci.data::expr_COPD_design</a></code>): -
<code>age_enroll</code>: age of the participant at the time of
enrollment in the study - <code>gender</code> - <code>bmi</code>: body
mass index - <code>parentalcopd</code>: parental history of COPD -
<code>smokcignow</code>: smoking status - <code>ats_packyears</code>:
Smoking history expressed in pack-years, calculated according to ATS
guidelines.</p>
<p>We also include an intercept in the design matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">X</span><span class="op">$</span><span class="va">intercept</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">X_copy</span> <span class="op">&lt;-</span> <span class="va">X</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">X_copy</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age_enroll"</span>, <span class="st">"ats_packyears"</span>, <span class="st">"bmi"</span>, <span class="st">"fev1_fvc_utah"</span>, </span>
<span>                     <span class="st">"gender"</span>, <span class="st">"parentalcopd"</span>, <span class="st">"smokcignow"</span>, <span class="st">"intercept"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The study aims at identifying associations between gene expression
and explanatory variables included in the design matrix. In the context
of a linear model, multiple hypotheses can be tested. For example, one
may test the difference between the two levels of the
<code>gender</code> variable, or assess the association between gene
expression and <code>fev1_fvc_utah</code>. This requires specifying a
contrast matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>
(of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">L \times p</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>L</mi><annotation encoding="application/x-tex">L</annotation></semantics></math>
is the number of contrasts and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
the number of genes) that encodes all hypotheses tested
simultaneously.</p>
</div>
<div class="section level2">
<h2 id="study-of-a-single-contrast">Study of a single contrast<a class="anchor" aria-label="anchor" href="#study-of-a-single-contrast"></a>
</h2>
<p>In this section, we only study the association between gene
expression and <code>fev1_fvc_utah</code> taking into account other
covariate in the model. We build the following contrast matrix:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tested_variable</span> <span class="op">&lt;-</span> <span class="st">"fev1_fvc_utah"</span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">*</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">==</span> <span class="va">tested_variable</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"association_lung_func"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">C</span></span>
<span><span class="co">#&gt;                       age_enroll ats_packyears bmi fev1_fvc_utah gender</span></span>
<span><span class="co">#&gt; association_lung_func          0             0   0             1      0</span></span>
<span><span class="co">#&gt;                       parentalcopd smokcignow intercept</span></span>
<span><span class="co">#&gt; association_lung_func            0          0         0</span></span></code></pre></div>
<p>For each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>,
we fit a model to estimate a vector of parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\beta_j</annotation></semantics></math>.
Then we test for each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
and for the contrast
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>,
if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>ℋ</mi><mn>0</mn></msub><mo>:</mo><mi>c</mi><msub><mi>β</mi><mi>j</mi></msub><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\mathcal{H}_0: c\beta_j = 0</annotation></semantics></math>.
In our case, we test if the parameter of the studied variable
<code>fev1_fvc_utah</code> is equal to 0. We want to obtain post hoc
guarantees on a set of genes which are associated with lung function
(contrast <code>fev1_fvc_utah</code>). We first build a
<code>SansSouci</code> class object given the gene expression matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>,
the design matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and the contrast matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>.
Then we fit the model and calibrate the post hoc bound using the method
<code><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit()</a></code>, setting the target risk to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.1</annotation></semantics></math>.
In the code below, we use only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><mo>=</mo><mn>100</mn></mrow><annotation encoding="application/x-tex">B = 100</annotation></semantics></math>
bootstraps to save computing time. In practice, we recommend taking
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><mo>=</mo><mn>1000</mn></mrow><annotation encoding="application/x-tex">B = 1000</annotation></semantics></math>
bootstraps for more stable results.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">Y</span>, X <span class="op">=</span> <span class="va">X</span>, Contrast <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="co"># B=100 to save computing time, but should use B=1000.</span></span></code></pre></div>
<p>For comparison purposes we also run the (parametric) Simes method
introduced by Goeman and Solari (2011):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span>, alpha <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span> <span class="co">## B=0 =&gt; no calibration!</span></span>
<span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"SansSouci"</span> <span class="op">=</span> <span class="va">res</span>,</span>
<span>                <span class="st">"Simes (parametric)"</span> <span class="op">=</span> <span class="va">res_Simes</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="confidence-curve">Confidence curve<a class="anchor" aria-label="anchor" href="#confidence-curve"></a>
</h3>
<p>In the absence of prior information on genes, a natural idea is to
rank them by decreasing statistical significance, and a natural question
to ask is:</p>
<blockquote>
<p>Can we provide a lower confidence curve on the number (or proportion)
of truly differentially expressed genes among the most significant
genes?</p>
</blockquote>
<p>The confidence curve obtained by calibration is the solid black line
in the figure below:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conf_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"darkgray"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">conf_bounds</span>, xmax <span class="op">=</span> <span class="fl">5000</span>, cols <span class="op">=</span> <span class="va">cols</span>, legend.title <span class="op">=</span> <span class="st">"Template"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/conf-curve-1-1.png" width="576"></p>
<p>This plot can be interpreted as follows: among the 2000 most
significant genes, the number of truly differentially expressed genes is
at least 1489 (right panel). Equivalently, the FDP among these 2000
genes is at most 0.26 (left panel).</p>
<p>The dashed gray curve is obtained by the parametric Simes method
introduced by Goeman and Solari (2011). The comparison between the two
curves illustrates the gain in power obtained by using permutation
methods to adapt to the dependence between genes. In this example, the
parametric Simes method gives the following guarantees: among the 2000
most significant genes, the number of truly differentially expressed
genes is at least 917 (right panel). Equivalently, the FDP among these
2000 genes is at most 0.54 (left panel).</p>
</div>
<div class="section level3">
<h3 id="differentially-expressed-genes">Differentially expressed genes<a class="anchor" aria-label="anchor" href="#differentially-expressed-genes"></a>
</h3>
<p>In this section we show how to the above curves may be used to
address the question:</p>
<blockquote>
<p>Which genes have their expression associated with lung function with
high probability?</p>
</blockquote>
<p>To do so, we <em>define</em> differentially expressed genes as the
largest set of genes for which the FDP bound is less than a user-given
value, for example
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>.
This corresponds to drawing a horizontal line in the preceding plot:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.1</span> <span class="co"># FDP budget (user-defined)</span></span>
<span><span class="va">FDP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, what <span class="op">=</span> <span class="st">"FDP"</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">n_DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">FDP</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">FDP</span>, xmax <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="va">n_DEG</span>, col <span class="op">=</span> <span class="va">cols</span>, legend.title <span class="op">=</span> <span class="st">"Template"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">n_DEG</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, col <span class="op">=</span> <span class="va">cols</span>, linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/post-hoc-FDP-control-1-1.png" width="576"></p>
<p>Using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>,
we obtain 1096 differentially expressed genes. Note that this gene list
has a clear statistical interpretation: with probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>α</mi><mo>=</mo><mn>0.9</mn></mrow><annotation encoding="application/x-tex">1-\alpha = 0.9</annotation></semantics></math>,
the proportion of false positives (that is, genes that are called DE by
mistake) is less than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>.</p>
<p>The above example also illustrates the increase in power obtained by
calibration, since the parametric Simes method yields a subset of “only”
546 genes called differentially expressed (with identical statistical
guarantees).</p>
</div>
<div class="section level3">
<h3 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h3>
<!-- For an interactive volcano plot, see the [volcano plot shiny application]( https://shiny-iidea-sanssouci.apps.math.cnrs.fr/). -->
<p>A classical practice in gene expression studies is to define DE genes
as those passing both a significance filter (small
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value)
and an effect size or “fold change” filter. Here, the fold change of a
gene is defined as the estimate of the tested contrast
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mover><mi>β</mi><mo accent="true">̂</mo></mover></mrow><annotation encoding="application/x-tex">c\hat\beta</annotation></semantics></math>.
This double selection by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
and fold change corresponds to two sets of genes, with positive/negative
fold change, which can be represented in the following plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, p <span class="op">=</span> <span class="fl">1e-3</span>, r <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/volcano-1-1.png" width="576"></p>
<p>This type of plot is called a “volcano plot” <span class="citation">Cui and Churchill (2003)</span>. Post hoc inference
makes it possible to obtain statistical guarantees on selections such as
the ones represented in the above figure.</p>
</div>
</div>
<div class="section level2">
<h2 id="test-several-contrasts-simultaneously">Test several contrasts simultaneously<a class="anchor" aria-label="anchor" href="#test-several-contrasts-simultaneously"></a>
</h2>
<p>If several contrasts are studied, to obtain post hoc guarantees, we
have to consider all tests (for all genes and all tested contrasts)
simultaneously at the inference step. The contrast matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>
contains all contrasts, one by row. For example, let us assume that we
we want to test the lung function variable <code>fev1_fvc_utah</code>
and the <code>age_enroll</code> variable. Then the contrast matrix is
then:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tested_variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age_enroll"</span>, <span class="st">"fev1_fvc_utah"</span><span class="op">)</span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tested_variables</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tested_variables</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">C</span></span>
<span><span class="co">#&gt;               age_enroll ats_packyears bmi fev1_fvc_utah gender parentalcopd</span></span>
<span><span class="co">#&gt; age_enroll             1             0   0             0      0            0</span></span>
<span><span class="co">#&gt; fev1_fvc_utah          0             0   0             1      0            0</span></span>
<span><span class="co">#&gt;               smokcignow intercept</span></span>
<span><span class="co">#&gt; age_enroll             0         0</span></span>
<span><span class="co">#&gt; fev1_fvc_utah          0         0</span></span></code></pre></div>
<p>We build another <code>SansSouci</code> object with this contrast
matrix, fit the model and calibrate the post hoc bounds.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">Y</span>, X <span class="op">=</span> <span class="va">X</span>, Contrast <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">object</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, B <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>Using information from all contrasts simultaneously is important to
ensure valid post hoc inference. However, the identification and
selection of differentially expressed (DE) genes are typically performed
contrast by contrast, in response to specific scientific questions. To
support this, the package provides visualization tools such as
confidence curves and volcano plots, which help interpret the results
for a given contrast.</p>
<div class="section level3">
<h3 id="confidence-curves">Confidence curves<a class="anchor" aria-label="anchor" href="#confidence-curves"></a>
</h3>
<p>By default, the confidence curves associated to each contrast are
displayed on the same plot, allowing for direct comparison of the signal
across multiple contrasts. In the example below, we can clearly observe
that the signal associated with the <code>fev1_fvc_utah</code> variable
is stronger than that of <code>age_enroll</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.1</span> <span class="co"># FDP budget (user-defined)</span></span>
<span><span class="va">FDP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res</span>, all <span class="op">=</span> <span class="cn">TRUE</span>, contrast_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age_enroll"</span>, <span class="st">"fev1_fvc_utah"</span><span class="op">)</span>, what <span class="op">=</span> <span class="st">"FDP"</span><span class="op">)</span></span>
<span><span class="va">n_DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">FDP</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span>, xmax <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="va">n_DEG</span>, contrast_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age_enroll"</span>, <span class="st">"fev1_fvc_utah"</span><span class="op">)</span>, what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FDP"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">n_DEG</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>,  linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/unnamed-chunk-2-1.png" width="576"></p>
<p>It is also possible to specify the contrast of interest. To do so for
the lung function (corresponding to the contrast
<code>fev1_fvc_utah</code>), we <em>define</em> differentially expressed
genes as the largest set of genes for which the FDP bound is less than a
user-given value, for example
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>.
This corresponds to drawing a horizontal line in the confidence curves
on the FDP:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.1</span> <span class="co"># FDP budget (user-defined)</span></span>
<span><span class="va">FDP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res</span>, all <span class="op">=</span> <span class="cn">TRUE</span>, contrast_name <span class="op">=</span> <span class="st">"fev1_fvc_utah"</span>, what <span class="op">=</span> <span class="st">"FDP"</span><span class="op">)</span></span>
<span><span class="va">n_DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">FDP</span><span class="op">$</span><span class="va">bound</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span></span>
<span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span>, xmax <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="va">n_DEG</span>, contrast_name <span class="op">=</span> <span class="st">"fev1_fvc_utah"</span>, what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FDP"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">n_DEG</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>,  linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/post-hoc-FDP-control-contrast-fev-1.png" width="576"></p>
<p>Using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>,
we obtain <code>r</code>n_DEG` differentially expressed genes. Note that
this gene list has a clear statistical interpretation: with probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>α</mi><mo>=</mo><mn>0.9</mn></mrow><annotation encoding="application/x-tex">1-\alpha = 0.9</annotation></semantics></math>,
the proportion of false positives (that is, genes that are called DE by
mistake) is less than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">q = 0.1</annotation></semantics></math>.</p>
<p>The same analysis can be made for the contrast on
<code>age_enroll</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.5</span> <span class="co"># FDP budget (user-defined)</span></span>
<span><span class="va">FDP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res</span>, all <span class="op">=</span> <span class="cn">TRUE</span>, contrast_name <span class="op">=</span> <span class="st">"age_enroll"</span>, what <span class="op">=</span> <span class="st">"FDP"</span><span class="op">)</span></span>
<span><span class="va">n_DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">FDP</span><span class="op">$</span><span class="va">bound</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span></span>
<span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span>, xmax <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="va">n_DEG</span>, contrast_name <span class="op">=</span> <span class="st">"age_enroll"</span>, what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FDP"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">n_DEG</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>,  linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="va">size</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/post-hoc-FDP-control-contrast-age-1.png" width="576"></p>
<p>Due to the weaker signal in this case, we had to increase the target
FDP level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>q</mi><annotation encoding="application/x-tex">q</annotation></semantics></math>
in order to obtain a relatively large set of genes selected.</p>
</div>
<div class="section level3">
<h3 id="selection-on-volcano-plot">Selection on Volcano Plot<a class="anchor" aria-label="anchor" href="#selection-on-volcano-plot"></a>
</h3>
<p>The volcano plot can be used to make selection of interest genes.
Here, the name of the contrast has to be specified explicilty.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, contrast_name <span class="op">=</span> <span class="st">"fev1_fvc_utah"</span>, p <span class="op">=</span> <span class="fl">1e-3</span>, r <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/volcano-contrast-fev-1.png" width="576"></p>
<p>Compared to the volcano plot obtained in the preceding section, the
selected genes are the same because the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
are identical. However, the confidence bounds are less tight. This is
explained by the fact that here two contrasts are tested instead of one:
the total number of hypotheses tested is twice as large as previously,
resulting in a more severe multiple testing correction.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, contrast_name <span class="op">=</span> <span class="st">"age_enroll"</span>, p <span class="op">=</span> <span class="fl">0.001</span>, r <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_linear-model_files/figure-html/volcano-contrast-age-1.png" width="576"></p>
<p>As noted above, the association of variable <code>age_enrol</code>
with gene expression is much weaker.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-bahr2013peripheral" class="csl-entry">
Bahr, Timothy M, Grant J Hughes, Michael Armstrong, Rick Reisdorph,
Christopher D Coldren, Michael G Edwards, Christina Schnell, et al.
2013. <span>“Peripheral Blood Mononuclear Cell Gene Expression in
Chronic Obstructive Pulmonary Disease.”</span> <em>American Journal of
Respiratory Cell and Molecular Biology</em> 49 (2): 316–23.
</div>
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-CC2003" class="csl-entry">
Cui, Xiangqin, and Gary A Churchill. 2003. <span>“Statistical Tests for
Differential Expression in c<span>DNA</span> Microarray
Experiments.”</span> <em>Genome Biol</em> 4 (4): 210.
</div>
<div id="ref-davenport2025fdp" class="csl-entry">
Davenport, Samuel, Bertrand Thirion, and Pierre Neuvial. 2025.
<span>“FDP Control in Mass-Univariate Linear Models Using the Residual
Bootstrap.”</span> <em>Electron. J. Statist.</em> 19 (1): 1313–36. <a href="https://doi.org/10.1214/25-EJS2354" class="external-link">https://doi.org/10.1214/25-EJS2354</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
