<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Permutation-based post hoc inference for two-group differential gene expression studies • sanssouci</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Permutation-based post hoc inference for two-group differential gene expression studies">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sanssouci</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.16.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/CopyOfconfidenceCurves_localized.html">Confidence curves for structured hypotheses</a></li>
    <li><a class="dropdown-item" href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a></li>
    <li><a class="dropdown-item" href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_linear-model.html">Bootstrap-based post hoc inference for differential gene expression studies in linear models</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for two-group differential gene expression studies: RNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for two-group differential gene expression studies</a></li>
    <li><a class="dropdown-item" href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a></li>
    <li><a class="dropdown-item" href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sanssouci-org/sanssouci/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Permutation-based post hoc inference for two-group differential gene expression studies</h1>
                        <h4 data-toc-skip class="author">Gilles
Blanchard, Pierre Neuvial and Etienne Roquain</h4>
            
            <h4 data-toc-skip class="date">2025-07-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sanssouci-org/sanssouci/blob/master/vignettes/post-hoc_differential-expression.Rmd" class="external-link"><code>vignettes/post-hoc_differential-expression.Rmd</code></a></small>
      <div class="d-none name"><code>post-hoc_differential-expression.Rmd</code></div>
    </div>

    
    
<p>This vignette reproduces figures from Section 9 in the book chapter:
<span class="citation">Blanchard, Neuvial, and Roquain (2021)</span>. It
demonstrates how the <a href="https://github.com/sanssouci-org/sanssouci" class="external-link"><code>sanssouci</code></a>
package may be used to obtain post hoc confidence bounds on false
positives in the case of differential gene expression analysis. After
showing the output of a classical differential analysis based on False
Discovery Rate control we illustrate the application of basic post-hoc
bounds derived from probabilistic inequalities. Then, we introduce more
powerful post hoc methods (introduced by <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span>) that
yield tighter bounds by adapting to unknown dependence by randomization.
Finally we demonstrate the use of these methods on two applications of
post hoc methods:</p>
<ul>
<li>build confidence curves (envelopes) for the true or false positives
and define differentially expressed genes accordingly</li>
<li>perform statistical inference on the output of volcano plots</li>
</ul>
<p>The methods described in this vignette are described in detail in the
book chapter <span class="citation">Blanchard, Neuvial, and Roquain
(2021)</span> and in the paper <span class="citation">Blanchard,
Neuvial, and Roquain (2020)</span>. A shiny application for volcano
plots is also available at <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link uri">https://shiny-iidea-sanssouci.apps.math.cnrs.fr/</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span> <span class="co"># install.packages("ggplot2")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span> <span class="co"># remotes::install_github("sanssouci-org/sanssouci@develop")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/github-org/sanssouci.data" class="external-link">"sanssouci.data"</a></span><span class="op">)</span> <span class="co"># remotes::install_github("sanssouci-org/sanssouci.data")</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sanssouci-org/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span></code></pre></div>
<p>Set the seed of the random number generator for numerical
reproducibility of the results:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">20200924</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="motivation-a-differential-gene-expression-study">Motivation: a differential gene expression study<a class="anchor" aria-label="anchor" href="#motivation-a-differential-gene-expression-study"></a>
</h2>
<p>We focus on differential gene expression studies in cancerology.
These studies aim at identifying genes whose mean expression level
differs significantly between two (or more) populations, based on a
sample of gene expression measurements from individuals from these
populations. Specifically, we consider a data set studied in <span class="citation">Bourgon, Gentleman, and Huber (2010)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">expr_ALL</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">expr_ALL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">expr_ALL</span><span class="op">)</span></span></code></pre></div>
<p>This data set consists of gene expression measurements for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>=</mo><mn>79</mn></mrow><annotation encoding="application/x-tex">n = 79</annotation></semantics></math>
patients with B-cell acute lymphoblastic leukemia (ALL) <span class="citation">Chiaretti et al. (2005)</span>. These patients are
classified into two subgoups, depending on whether or not they harbor a
specific mutation called “BCR/ABL”:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; BCR/ABL     NEG </span></span>
<span><span class="co">#&gt;      37      42</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></code></pre></div>
<p>The goal of this study is to understand the molecular differences at
the gene expression level between the populations of BCR/ABL positive
and negative (“NEG”) patients. For each patient, we observe a vector of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>=</mo><mn>9038</mn></mrow><annotation encoding="application/x-tex">m = 9038</annotation></semantics></math>
gene expression values.</p>
<p>The most basic question to ask is:</p>
<blockquote>
<p>For which genes is there a difference in the mean expression level of
the mutated and non-mutated population?</p>
</blockquote>
<p>This question can be addressed by performing one statistical test of
no difference between means for each gene, and to define “differentially
expressed” genes as those passing some significance threshold.</p>
<p>Below, the Welch test for differential expression is applied to each
gene. This can be done e.g. using the
<code><a href="../reference/rowWelchTests.html">sanssouci::rowWelchTests</a></code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="op">==</span> <span class="st">"BCR/ABL"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span> <span class="co"># map to 0/1</span></span>
<span><span class="va">dex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="../reference/rowWelchTests.html">rowWelchTests</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">categ</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pval</span> <span class="op">&lt;-</span> <span class="va">dex</span><span class="op">[[</span><span class="st">"p.value"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We plot a histogram of the corresponding
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">pval</span>, probability <span class="op">=</span> <span class="cn">TRUE</span>, breaks <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"p-value"</span>, main <span class="op">=</span> <span class="st">"p-value distributon"</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/hist-1.png" width="768"></p>
<p>As expected, the distribution presents a large number of small
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
(which include signals, i.e. differentially expressed genes) mixed with
uniformly distributed
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
(corresponding to non-differentially expressed genes).</p>
<div class="section level3">
<h3 id="multiple-testing-correction-false-discovery-rate-control">Multiple testing correction: False Discovery Rate control<a class="anchor" aria-label="anchor" href="#multiple-testing-correction-false-discovery-rate-control"></a>
</h3>
<p>The state of the art approach to large-scale multiple testing is to
control the False Discovery Rate (FDR), which is the expected proportion
of wrongly selected genes (false positives) among all selected genes
<span class="citation">Benjamini and Hochberg (1995)</span>. The most
widely used method to control this risk is the Benjamini-Hochberg (BH)
procedure, which has been shown to control the FDR when the hypotheses
corresponding to the non-differentially expressed genes are independent
<span class="citation">Benjamini and Hochberg (1995)</span> or satisfy a
specific type of positive dependence called Positive Regression
Dependence on the Subset (PRDS)
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ℋ</mi><mn>0</mn></msub><annotation encoding="application/x-tex">\mathcal{H}_0</annotation></semantics></math>
of truly non-differentially expressed genes <span class="citation">Benjamini and Yekutieli (2001)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">adjp_BH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">dex</span><span class="op">$</span><span class="va">adjp</span> <span class="op">&lt;-</span> <span class="va">adjp_BH</span></span>
<span><span class="va">S_BH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">adjp_BH</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span></span>
<span><span class="va">nBH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">S_BH</span><span class="op">)</span></span>
<span><span class="va">nBH</span></span>
<span><span class="co">#&gt; [1] 150</span></span></code></pre></div>
<p>The application of the BH procedure at level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">q = 0.05</annotation></semantics></math>
is illustrated in the figures below (all genes are displayed in the
first one, second one is a zoom on the top genes):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_col</span> <span class="op">&lt;-</span> <span class="st">"#FF000080"</span></span>
<span><span class="va">dexo</span> <span class="op">&lt;-</span> <span class="va">dex</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>, <span class="op">]</span>  <span class="co">## order genes by increasing p-values</span></span>
<span><span class="va">dexo</span><span class="op">[[</span><span class="st">"gene_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bh_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dexo</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gene_order</span>, y <span class="op">=</span> <span class="va">p.value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Number of top genes"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Ordered p-value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">m</span>, intercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="va">q</span><span class="op">/</span><span class="va">m</span>, color <span class="op">=</span> <span class="va">my_col</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># geom_segment(aes(x = nBH, y = 0, yend = q*nBH/m, xend = nBH), linetype = "dotted") +</span></span>
<span>  <span class="co"># geom_segment(aes(x = 0, y = q*nBH/m, xend = nBH, yend = q*nBH/m), linetype = "dotted") +</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">0</span>, intercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, color <span class="op">=</span> <span class="va">my_col</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span>, </span>
<span>        axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">18</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#geom_text(x = 0, y = q, label = expression(alpha), color = my_col)</span></span>
<span></span>
<span><span class="va">bh_plot</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/bh-plot-1.png" width="768"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xmax</span> <span class="op">&lt;-</span> <span class="va">nBH</span><span class="op">*</span><span class="fl">2.5</span></span>
<span><span class="va">ymax</span> <span class="op">&lt;-</span> <span class="va">dexo</span><span class="op">$</span><span class="va">p.value</span><span class="op">[</span><span class="va">xmax</span><span class="op">]</span></span>
<span><span class="va">bh_plot</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">xmax</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">ymax</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nBH</span>, y <span class="op">=</span> <span class="fl">0</span>, yend <span class="op">=</span> <span class="va">q</span><span class="op">*</span><span class="va">nBH</span><span class="op">/</span><span class="va">m</span>, xend <span class="op">=</span> <span class="va">nBH</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span>, y <span class="op">=</span> <span class="va">q</span><span class="op">*</span><span class="va">nBH</span><span class="op">/</span><span class="va">m</span>, xend <span class="op">=</span> <span class="va">nBH</span>, yend <span class="op">=</span> <span class="va">q</span><span class="op">*</span><span class="va">nBH</span><span class="op">/</span><span class="va">m</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, col <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in geom_segment(aes(x = nBH, y = 0, yend = q * nBH/m, xend = nBH), : All aesthetics have length 1, but the data has 9038 rows.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please consider using `annotate()` or provide this layer with data containing</span></span>
<span><span class="co">#&gt;   a single row.</span></span>
<span><span class="co">#&gt; Warning in geom_segment(aes(x = 1, y = q * nBH/m, xend = nBH, yend = q * : All aesthetics have length 1, but the data has 9038 rows.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please consider using `annotate()` or provide this layer with data containing</span></span>
<span><span class="co">#&gt;   a single row.</span></span>
<span><span class="co">#&gt; Warning: Removed 8663 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/bh-plot-zoom-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="basic-posthoc-bounds">Basic posthoc bounds<a class="anchor" aria-label="anchor" href="#basic-posthoc-bounds"></a>
</h3>
<p>Post hoc inference makes it possible to build confidence statements
on the number of true/false positives within any set
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
of genes:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
may be selected after seing the data (e.g.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
may be the set of rejections by the BH prcedure), and multiple choices
of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
are allowed. Post hoc inference has been popularized by <span class="citation">Goeman and Solari (2011)</span>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">pvalo</span> <span class="op">&lt;-</span> <span class="va">dexo</span><span class="op">$</span><span class="va">p.value</span></span></code></pre></div>
<div class="section level4">
<h4 id="k_0-bonferroni-bound">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mn>0</mn></msub><annotation encoding="application/x-tex">k_0</annotation></semantics></math>-Bonferroni
bound<a class="anchor" aria-label="anchor" href="#k_0-bonferroni-bound"></a>
</h4>
<p>For a fixed
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mn>0</mn></msub><annotation encoding="application/x-tex">k_0</annotation></semantics></math>,
the generalized Bonferroni procedure consisting in rejecting the
hypotheses in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>R</mi><msub><mi>k</mi><mn>0</mn></msub></msub><mo>=</mo><mo stretchy="false" form="prefix">{</mo><mi>i</mi><mo>:</mo><msub><mi>p</mi><mi>i</mi></msub><mo>≤</mo><mi>α</mi><msub><mi>k</mi><mn>0</mn></msub><mi>/</mi><mi>m</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">R_{k_0} = \{i: p_i \leq \alpha k_0/m\}</annotation></semantics></math>
controls the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mn>0</mn></msub><annotation encoding="application/x-tex">k_0</annotation></semantics></math>-Family-Wise
Error Rate: it ensures that with probability larger than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>α</mi></mrow><annotation encoding="application/x-tex">1-\alpha</annotation></semantics></math>,
the number of false positives in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>R</mi><msub><mi>k</mi><mn>0</mn></msub></msub><annotation encoding="application/x-tex">R_{k_0}</annotation></semantics></math>
is not larger than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>k</mi><mn>0</mn></msub><mo>−</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">k_0-1</annotation></semantics></math>.
As noted in <span class="citation">Blanchard, Neuvial, and Roquain
(2021)</span>, this leads to the post hoc bound:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>S</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><munder><mo>∑</mo><mrow><mi>i</mi><mo>∈</mo><mi>S</mi></mrow></munder><msub><mn>1</mn><mrow><mo stretchy="false" form="prefix">{</mo><msub><mi>p</mi><mi>i</mi></msub><mo>≥</mo><mi>α</mi><msub><mi>k</mi><mn>0</mn></msub><mi>/</mi><mrow><mo stretchy="true" form="prefix">|</mo><mi>S</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="false" form="postfix">}</mo></mrow></msub><mo>+</mo><msub><mi>k</mi><mn>0</mn></msub><mo>−</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">V(S)=\sum_{i\in S} 1_{\{p_i \geq \alpha k_0 /|S|\}} + k_0-1</annotation></semantics></math></p>
<p>As an application we calculate the bound associated to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>k</mi><mn>0</mn></msub><mo>=</mo><mn>100</mn></mrow><annotation encoding="application/x-tex">k_0=100</annotation></semantics></math>
for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0.1</mn></mrow><annotation encoding="application/x-tex">\alpha = 0.1</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k0</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">S</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="va">nBH</span></span>
<span><span class="va">FP_k0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pvalo</span><span class="op">[</span><span class="va">S</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="va">alpha</span><span class="op">*</span><span class="va">k0</span><span class="op">/</span><span class="va">m</span><span class="op">)</span> <span class="op">+</span> <span class="va">k0</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">FP_k0</span><span class="op">/</span><span class="va">nBH</span></span>
<span><span class="co">#&gt; [1] 0.66</span></span></code></pre></div>
<p>This implies that with probability larger than 0.9 the false
discovery proportion among the genes selected by the BH procedure at
level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">q= 0.05</annotation></semantics></math>
is upper bounded by 0.66.</p>
</div>
<div class="section level4">
<h4 id="simes-bound">Simes bound<a class="anchor" aria-label="anchor" href="#simes-bound"></a>
</h4>
<p>A more refined post hoc bound has been proposed by <span class="citation">Goeman and Solari (2011)</span> under the PRDS
assumption. In the framework of <span class="citation">Blanchard,
Neuvial, and Roquain (2021)</span> this bound is a direct consequence of
the <span class="citation">Simes (1986)</span> inequality. It can be
applied to the 150 rejections of the BH procedure as follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">dat</span>, groups <span class="op">=</span> <span class="va">categ</span><span class="op">)</span></span>
<span><span class="va">res_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span>, alpha <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span> <span class="co">## B=0 =&gt; no calibration!</span></span>
<span><span class="va">FP_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">res_Simes</span>, <span class="va">S_BH</span>, what <span class="op">=</span> <span class="st">"FP"</span><span class="op">)</span></span></code></pre></div>
<p>The Simes bound implies that with probability larger than 0.9, the
false discovery proportion among the genes selected by the BH procedure
at level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">q= 0.05</annotation></semantics></math>
is upper bounded by 0.44. The Simes bound is sharper than the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mn>0</mn></msub><annotation encoding="application/x-tex">k_0</annotation></semantics></math>-Bonferroni
bound because it is obtained from a <em>joint</em> control of all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-FWER
for all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>…</mi><mo>,</mo><mi>m</mi></mrow><annotation encoding="application/x-tex">k=1, \dots, m</annotation></semantics></math>.
The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>k</mi><mn>0</mn></msub><annotation encoding="application/x-tex">k_0</annotation></semantics></math>-Bonferroni
bound will therefore not be considered further in this vignette.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="tighter-confidence-bounds-by-adaptation-to-unknown-dependence">Tighter confidence bounds by adaptation to unknown dependence<a class="anchor" aria-label="anchor" href="#tighter-confidence-bounds-by-adaptation-to-unknown-dependence"></a>
</h2>
<p>As discussed in <span class="citation">Blanchard, Neuvial, and
Roquain (2020)</span>, the above-described bound has two major
limitations, being a consequence of the Simes inequality:</p>
<ul>
<li><p>It is known to be valid only under certain positive dependence
assumptions (PRDS) on the joint
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
distribution. Although the PRDS assumption is generally accepted in the
case of differential expression studies (which justifies the application
of the BH procedure itself), it has not been formally proved to hold in
this case.</p></li>
<li><p>It is not <em>adaptive</em> to the specific type of dependence at
hand for a particular data set.</p></li>
</ul>
<p>To bypass these limitations, <span class="citation">Blanchard,
Neuvial, and Roquain (2020)</span> have proposed a randomization-based
procedure known as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>-calibration,
which yields tighter bounds that are adapted to the dependency observed
in the data set at hand. A closely related approach has been proposed by
<span class="citation">Hemerik, Solari, and Goeman (2019)</span>. In the
case of two-sample tests, this calibration can be achieved by
permutation of class labels:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="va">B</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, family <span class="op">=</span> <span class="st">"Simes"</span><span class="op">)</span></span></code></pre></div>
<p>An alternative to the Simes/Linear reference family is the Beta
reference family:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">res_Beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">res</span>, B <span class="op">=</span> <span class="va">B</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, family <span class="op">=</span> <span class="st">"Beta"</span>, K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span></code></pre></div>
<p>As expected from the theory, the post hoc bounds obtained after
calibration by these methods is much tighter than the Simes bound:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Simes"</span> <span class="op">=</span> <span class="va">res_Simes</span>,</span>
<span>            <span class="st">"Linear"</span> <span class="op">=</span> <span class="va">res</span>,</span>
<span>            <span class="st">"Beta"</span> <span class="op">=</span> <span class="va">res_Beta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">resList</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Beta (K=%s)"</span>, <span class="va">K</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, <span class="va">S_BH</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bounds</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lower bound on True Positives"</span>, <span class="st">"Upper bound on False Discovery Proportion"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">bounds</span><span class="op">)</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="35%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">Lower bound on True Positives</th>
<th align="right">Upper bound on False Discovery Proportion</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Simes</td>
<td align="right">85</td>
<td align="right">0.43</td>
</tr>
<tr class="even">
<td align="left">Linear</td>
<td align="right">122</td>
<td align="right">0.19</td>
</tr>
<tr class="odd">
<td align="left">Beta (K=50)</td>
<td align="right">132</td>
<td align="right">0.12</td>
</tr>
</tbody>
</table>
<p>In the next two sections we illustrate the use of these improved
bounds in order to build</p>
<ul>
<li>confidence curves for the true or false positives</li>
<li>confidence statements for volcano plots</li>
</ul>
</div>
<div class="section level2">
<h2 id="confidence-curves-on-top-k-lists">Confidence curves on
“top-<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>”
lists<a class="anchor" aria-label="anchor" href="#confidence-curves-on-top-k-lists"></a>
</h2>
<p>In the absence of prior information on genes, a natural idea is to
rank them by decreasing statistical significance, and a natural question
to ask is:</p>
<blockquote>
<p>Can we provide a lower confidence curve on the number (or proportion)
of truly differentially expressed genes among the most significant
genes?</p>
</blockquote>
<p>We illustrate the use of post-hoc methods to provide this type of
information. More specifcally, we build confidence statements on the
number of true/false positives within the top
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
most significant genes in a differential gene expression study, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
may be defined by the user after seing the data, and multiple choices of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
are allowed.</p>
<p>The confidence curves obtained by calibration of the Simes and Beta
families can be compared graphically to the (parametric) Simes curve
that can be obtained from <span class="citation">Goeman and Solari
(2011)</span>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conf_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># cols &lt;- c("lightgray", "black", "darkgray")</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conf_bounds</span><span class="op">)</span>, <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">conf_bounds</span>, xmax <span class="op">=</span> <span class="fl">300</span>, cols <span class="op">=</span> <span class="va">cols</span>, legend.title <span class="op">=</span> <span class="st">"Template"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">nBH</span>, color <span class="op">=</span> <span class="st">"gray"</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/conf-env-plot-1.png" width="768"></p>
<p>Both calibrated curves outperform the Simes curve in this
example.</p>
</div>
<div class="section level2">
<h2 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h2>
<p>For an interactive volcano plot, see the <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link">volcano plot
shiny application</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span>  <span class="fl">0.3</span></span></code></pre></div>
<p>Let us assume that we are interested in genes selected by the BH
procedure at level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>q</mi><mo>=</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">q = 0.05</annotation></semantics></math>
and whose fold change is larger than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mo>=</mo><mn>0.3</mn></mrow><annotation encoding="application/x-tex">r = 0.3</annotation></semantics></math>
in absolute value. The “fold change” is defined as the difference
between the expression means of the two groups compared; it is an
estimate of the effect size of a gene. This double selection corresponds
to two sets of genes, with positive/negative fold change, which can be
represented in the following plot:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ylim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res_Simes</span>, q <span class="op">=</span> <span class="va">q</span>, r <span class="op">=</span> <span class="va">r</span>, ylim <span class="op">=</span> <span class="va">ylim</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/volcano-Simes-1.png" width="768"></p>
<p>This type of plot is called a “volcano plot” <span class="citation">Cui and Churchill (2003)</span>. Post hoc inference
makes it possible to obtain statistical guarantees on selections such as
the ones represented in the above figure.</p>
<p>The substantial gain in power offered by the above-described
calibration is illustrated as follows for the Simes reference
family:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, q <span class="op">=</span> <span class="va">q</span>, r <span class="op">=</span> <span class="va">r</span>, ylim <span class="op">=</span> <span class="va">ylim</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/volcano-Simes-cal-1.png" width="768"></p>
<p>and for the Beta reference family.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res_Beta</span>, q <span class="op">=</span> <span class="va">q</span>, r <span class="op">=</span> <span class="va">r</span>, ylim <span class="op">=</span> <span class="va">ylim</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_files/figure-html/volcano-Beta-cal-1.png" width="768"></p>
<p>The comparison between these bounds may be summarized by the
following Table:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/all-generics.html">foldChanges</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">S_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">fc</span> <span class="op">&gt;=</span> <span class="va">r</span> <span class="op">&amp;</span> <span class="va">adjp_BH</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span></span>
<span><span class="va">S_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">fc</span> <span class="op">&lt;=</span> <span class="op">-</span><span class="va">r</span> <span class="op">&amp;</span> <span class="va">adjp_BH</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span></span>
<span><span class="va">S_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">S_pos</span>, <span class="va">S_neg</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_bounds</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">S</span>, <span class="va">resList</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">S</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, <span class="va">S</span>, <span class="st">"TP"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu">all_bounds</span><span class="op">(</span><span class="va">S_pos</span>, <span class="va">resList</span><span class="op">)</span>, </span>
<span>             <span class="fu">all_bounds</span><span class="op">(</span><span class="va">S_neg</span>, <span class="va">resList</span><span class="op">)</span>,</span>
<span>             <span class="fu">all_bounds</span><span class="op">(</span><span class="va">S_all</span>, <span class="va">resList</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"BH-adjusted p.value &lt;"</span>, <span class="va">q</span><span class="op">)</span></span>
<span><span class="va">lab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">plab</span>, <span class="st">"&amp;"</span>, <span class="st">" fold change  &gt; "</span>, <span class="va">r</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">plab</span>, <span class="st">"&amp;"</span>, <span class="st">" fold change  &lt; "</span>, <span class="op">-</span><span class="va">r</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">plab</span>, <span class="st">"&amp;"</span>, <span class="st">"|fold change| &gt; "</span>, <span class="va">r</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">lab</span>, <span class="va">tab</span><span class="op">)</span></span>
<span><span class="va">cap</span> <span class="op">&lt;-</span> <span class="st">"Post hoc bounds on true positives in user-defined gene selections"</span></span>
<span><span class="co">#knitr::kable(tab, caption = cap, format = "latex")</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">tab</span>, caption <span class="op">=</span> <span class="va">cap</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<caption>Post hoc bounds on true positives in user-defined gene
selections</caption>
<colgroup>
<col width="67%">
<col width="4%">
<col width="6%">
<col width="7%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="left">lab</th>
<th align="left"></th>
<th align="left">Simes</th>
<th align="left">Linear</th>
<th align="left">Beta (K=50)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BH-adjusted p.value &lt; 0.05 &amp; fold change &gt;
0.3</td>
<td align="left">111</td>
<td align="left">60</td>
<td align="left">86</td>
<td align="left">93</td>
</tr>
<tr class="even">
<td align="left">BH-adjusted p.value &lt; 0.05 &amp; fold change &lt;
-0.3</td>
<td align="left">22</td>
<td align="left">1</td>
<td align="left">7</td>
<td align="left">6</td>
</tr>
<tr class="odd">
<td align="left">BH-adjusted p.value &lt; 0.05 &amp; |fold change| &gt;
0.3</td>
<td align="left">133</td>
<td align="left">74</td>
<td align="left">107</td>
<td align="left">115</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] sanssouci_0.16.0 ggplot2_3.5.2   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.7-3       gtable_0.3.6       jsonlite_2.0.0     compiler_4.5.1    </span></span>
<span><span class="co">#&gt;  [5] crayon_1.5.3       jquerylib_0.1.4    systemfonts_1.2.3  scales_1.4.0      </span></span>
<span><span class="co">#&gt;  [9] textshaping_1.0.1  yaml_2.3.10        fastmap_1.2.0      lattice_0.22-7    </span></span>
<span><span class="co">#&gt; [13] R6_2.6.1           labeling_0.4.3     generics_0.1.4     knitr_1.50        </span></span>
<span><span class="co">#&gt; [17] tibble_3.3.0       desc_1.4.3         bslib_0.9.0        pillar_1.11.0     </span></span>
<span><span class="co">#&gt; [21] RColorBrewer_1.1-3 rlang_1.1.6        cachem_1.1.0       xfun_0.52         </span></span>
<span><span class="co">#&gt; [25] fs_1.6.6           sass_0.4.10        cli_3.6.5          pkgdown_2.1.3     </span></span>
<span><span class="co">#&gt; [29] withr_3.0.2        magrittr_2.0.3     matrixTests_0.2.3  digest_0.6.37     </span></span>
<span><span class="co">#&gt; [33] grid_4.5.1         lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.4    </span></span>
<span><span class="co">#&gt; [37] glue_1.8.0         farver_2.1.2       codetools_0.2-20   ragg_1.4.0        </span></span>
<span><span class="co">#&gt; [41] rmarkdown_2.29     matrixStats_1.5.0  tools_4.5.1        pkgconfig_2.0.3   </span></span>
<span><span class="co">#&gt; [45] htmltools_0.5.8.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>To re-build this vignette from its source, use:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_differential-expression.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"pdf_document"</span><span class="op">)</span></span>
<span><span class="co"># To keep intermediate files, add option 'clean = FALSE'</span></span>
<span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_differential-expression.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"html_document"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-benjamini95controlling" class="csl-entry">
Benjamini, Yoav, and Yosef Hochberg. 1995. <span>“Controlling the False
Discovery Rate: A Practical and Powerful Approach to Multiple
Testing.”</span> <em>Journal of the Royal Statistical Society: Series B
(Methodological)</em> 57 (1): 289–300.
</div>
<div id="ref-benjamini01control" class="csl-entry">
Benjamini, Yoav, and Daniel Yekutieli. 2001. <span>“The Control of the
False Discovery Rate in Multiple Testing Under Dependency.”</span>
<em>Annals of Statistics</em>, 1165–88.
</div>
<div id="ref-blanchard20post-hoc" class="csl-entry">
Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020.
<span>“Post Hoc Confidence Bounds on False Positives Using Reference
Families.”</span> <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.
</div>
<div id="ref-BNR:chap" class="csl-entry">
———. 2021. <span>“<span class="nocase">On agnostic post hoc approaches
to false positive control</span>.”</span> In <em><span class="nocase">Handbook of Multiple Comparisons</span></em>, edited by
Xinping Cui, Thorsten Dickhaus, Ying Ding, and Jason C. Hsu. Handbooks
of Modern Statistical Methods. <span>Chapman<br>
&amp; Hall/CRC</span>. <a href="https://hal.archives-ouvertes.fr/hal-02320543" class="external-link">https://hal.archives-ouvertes.fr/hal-02320543</a>.
</div>
<div id="ref-BGH2010" class="csl-entry">
Bourgon, Richard, Robert Gentleman, and Wolfgang Huber. 2010.
<span>“<span class="nocase">Independent filtering increases detection
power for high-throughput experiments</span>.”</span> <em>PNAS</em>. <a href="https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107" class="external-link">https://doi.org/10.1073/pnas.0914005107/-/DCSupplemental.www.pnas.org/cgi/doi/10.1073/pnas.0914005107</a>.
</div>
<div id="ref-CLGV+2005" class="csl-entry">
Chiaretti, Sabina, Xiaochun Li, Robert Gentleman, Antonella Vitale,
Kathy S Wang, Franco Mandelli, Robin Foa, and Jerome Ritz. 2005.
<span>“Gene Expression Profiles of b-Lineage Adult Acute Lymphocytic
Leukemia Reveal Genetic Patterns That Identify Lineage Derivation and
Distinct Mechanisms of Transformation.”</span> <em>Clinical Cancer
Research</em> 11 (20): 7209–19.
</div>
<div id="ref-CC2003" class="csl-entry">
Cui, Xiangqin, and Gary A Churchill. 2003. <span>“Statistical Tests for
Differential Expression in c<span>DNA</span> Microarray
Experiments.”</span> <em>Genome Biol</em> 4 (4): 210.
</div>
<div id="ref-GS2011" class="csl-entry">
Goeman, Jelle J, and Aldo Solari. 2011. <span>“Multiple Testing for
Exploratory Research.”</span> <em>Statistical Science</em> 26 (4):
584–97.
</div>
<div id="ref-HSG2019" class="csl-entry">
Hemerik, Jesse, Aldo Solari, and Jelle J Goeman. 2019.
<span>“Permutation-Based Simultaneous Confidence Bounds for the False
Discovery Proportion.”</span> <em>Biometrika</em> 106 (3): 635–49.
</div>
<div id="ref-simes86improved" class="csl-entry">
Simes, R J. 1986. <span>“<span class="nocase">An improved Bonferroni
procedure for multiple tests of significance</span>.”</span>
<em>Biometrika</em> 73 (3): 751–54.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre Neuvial, Gilles Blanchard, Guillermo Durand, Nicolas Enjalbert-Courrech, Etienne Roquain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
