<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Permutation-based post hoc inference for differential gene expression studies: RNAseq data • sanssouci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Permutation-based post hoc inference for differential gene expression studies: RNAseq data">
<meta property="og:description" content="sanssouci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sanssouci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/IIDEA.html">IIDEA: Interactive Inference for Differential Expression Analyses</a>
    </li>
    <li>
      <a href="../articles/Simes_equi-correlation.html">Conservativeness of the Simes inequality under positive dependence</a>
    </li>
    <li>
      <a href="../articles/confidenceCurves_localized.html">Confidence curves for structured hypotheses</a>
    </li>
    <li>
      <a href="../articles/jointErrorRateCalibration_simulations.html">Joint Error Rate calibration</a>
    </li>
    <li>
      <a href="../articles/post-hoc_differential-expression.html">Permutation-based post hoc inference for differential gene expression studies</a>
    </li>
    <li>
      <a href="../articles/post-hoc_differential-expression_RNAseq.html">Permutation-based post hoc inference for differential gene expression studies: RNAseq data</a>
    </li>
    <li>
      <a href="../articles/post-hoc_fMRI.html">Permutation-based post hoc inference for fMRI studies</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/sanssouci/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="post-hoc_differential-expression_RNAseq_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Permutation-based post hoc inference for differential gene expression studies: RNAseq data</h1>
                        <h4 data-toc-skip class="author">Nicolas Enjalbert Courrech and Pierre Neuvial</h4>
            
            <h4 data-toc-skip class="date">2022-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pneuvial/sanssouci/blob/HEAD/vignettes/post-hoc_differential-expression_RNAseq.Rmd" class="external-link"><code>vignettes/post-hoc_differential-expression_RNAseq.Rmd</code></a></small>
      <div class="hidden name"><code>post-hoc_differential-expression_RNAseq.Rmd</code></div>

    </div>

    
    
<p>This vignette illustrates the relevance of permutation-based post hoc bounds on false positives for the differential analysis of (bulk) RNA sequencing (RNAseq) data. We illustrate how the post hoc methods introduced by <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span> may be used to</p>
<ul>
<li>build confidence curves (envelopes) for the true or false positives and defined differentially expressed genes accordingly</li>
<li>perform statistical inference on the output of volcano plots</li>
</ul>
<p>The methods described in this vignette have been introduced in the paper <span class="citation">Blanchard, Neuvial, and Roquain (2020)</span> and in the book chapter <span class="citation">Blanchard, Neuvial, and Roquain (2021)</span>. A shiny application for volcano plots is also available from <a href="https://shiny-iidea-sanssouci.apps.math.cnrs.fr/" class="external-link uri">https://shiny-iidea-sanssouci.apps.math.cnrs.fr/</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/pneuvial/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span> <span class="op">||</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pneuvial/sanssouci@develop"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/pneuvial/sanssouci" class="external-link">"sanssouci"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">"matrixStats"</a></span><span class="op">)</span></span></code></pre></div>
<p>Set the seed of the random number generator for numerical reproducibility of the results:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">20200924</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="motivation-a-differential-gene-expression-study">Motivation: a differential gene expression study<a class="anchor" aria-label="anchor" href="#motivation-a-differential-gene-expression-study"></a>
</h2>
<p>We focus on differential gene expression studies in cancerology. These studies aim at identifying genes whose mean expression level differs significantly between two (or more) populations, based on a sample of gene expression measurements from individuals from these populations.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"RNAseq_blca"</span>, package <span class="op">=</span> <span class="st">"sanssouci.data"</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">RNAseq_blca</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">RNAseq_blca</span><span class="op">)</span> <span class="op">==</span> <span class="st">"III"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">RNAseq_blca</span><span class="op">)</span></span></code></pre></div>
<p>This data set consists of gene expression measurements for <span class="math inline">\(n = 270\)</span> patients from the Cancer Genome Atlas Urothelial Bladder Carcinoma (TCGA-BLCA) data collection (<span class="citation">Cancer Genome Atlas Research Network and others (2014)</span>). Following a standard practice in RNAseq data analysis, we convert these measurements to to <em>count per millions</em> (CPM) in order to make them comparable across samples, and filter out genes with very low counts in all samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CPM</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">*</span><span class="fl">1e6</span></span>
<span><span class="co"># plot(density(rowMaxs(log(1 + CPM))))</span></span>
<span><span class="va">ww</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowRanges.html" class="external-link">rowMaxs</a></span><span class="op">(</span><span class="va">CPM</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">ww</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowQuantiles.html" class="external-link">rowQuantiles</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">CPM</span><span class="op">)</span>, prob <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">CPM</span><span class="op">[</span><span class="op">-</span><span class="va">ww</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html" class="external-link">density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">Y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_RNAseq_files/figure-html/CPM+filter-1.png" width="768"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html" class="external-link">density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">CPM</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_RNAseq_files/figure-html/CPM+filter-2.png" width="768"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<p>The patients are classified into two subgoups, according to their clinical stage (II or III):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span></span>
<span><span class="co">#&gt; groups</span></span>
<span><span class="co">#&gt;   0   1 </span></span>
<span><span class="co">#&gt; 130 140</span></span></code></pre></div>
<p>The goal of this study is to understand the molecular differences at the gene expression level between the stage II and stage III populations. Each patient is characterized by a vector of <span class="math inline">\(m = 12534\)</span> gene expression values. We aim at addressing the following question:</p>
<blockquote>
<p>For which genes is there a difference in the distribution of expression levels between the stage II and stage III populations?</p>
</blockquote>
<p>This question can be addressed by performing one statistical test per gene, and to define “differentially expressed” genes as those passing some significance threshold. Below, we use the Wilcoxon rank sum (or Mann-Whitney) test for comparing two independent samples. This is a rank-based non-parametric test. A fast version of this test is implemented in the <code><a href="../reference/rowWilcoxonTests.html">sanssouci::rowWilcoxonTests</a></code> function.</p>
<p>To obtain post hoc bounds, we create an object of class <code>SansSouci</code> and run the <code>fit</code> method to perform calibration with <span class="math inline">\(B=1,000\)</span> permutations. The target risk level is set to <span class="math inline">\(\alpha = 0.1\)</span>, meaning that all of the statements made below will be true with probability greater than <span class="math inline">\(1-\alpha= 0.9\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SansSouci.html">SansSouci</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">Y</span><span class="op">)</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="fl">1000</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, family <span class="op">=</span> <span class="st">"Simes"</span>, </span>
<span>           rowTestFUN <span class="op">=</span> <span class="va">rowWilcoxonTests</span><span class="op">)</span></span></code></pre></div>
<p>For comparison purposes we also run the (parametric) Simes method introduced by <span class="citation">Goeman and Solari (2011)</span>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_Simes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html" class="external-link">fit</a></span><span class="op">(</span><span class="va">obj</span>, B <span class="op">=</span> <span class="fl">0</span>, family <span class="op">=</span> <span class="st">"Simes"</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, </span>
<span>                 rowTestFUN <span class="op">=</span> <span class="va">rowWilcoxonTests</span><span class="op">)</span> <span class="co">## B=0 =&gt; no calibration!</span></span>
<span></span>
<span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"SansSouci"</span> <span class="op">=</span> <span class="va">res</span>,</span>
<span>                <span class="st">"Simes (parametric)"</span> <span class="op">=</span> <span class="va">res_Simes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="confidence-curves-for-top-k-lists">Confidence curves for “top-<span class="math inline">\(k\)</span>” lists<a class="anchor" aria-label="anchor" href="#confidence-curves-for-top-k-lists"></a>
</h2>
<p>In the absence of prior information on genes, a natural idea is to rank them by decreasing statistical significance, and a natural question to ask is:</p>
<blockquote>
<p>Can we provide a lower confidence curve on the number (or proportion) of truly differentially expressed genes among the most significant genes?</p>
</blockquote>
<p>The confidence curve obtained by calibration is the solid black line in the figure below:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conf_bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"darkgray"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">conf_bounds</span>, xmax <span class="op">=</span> <span class="fl">5000</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="co"># ggsave(p, file = "conf-curve.pdf", width = 6, height = 4)</span></span></code></pre></div>
<p>This plot can be interpreted as follows: among the 2000 most significant genes, the number of truly differentially expressed genes is at least 1163 (right panel). Equivalently, the FDP among these 2000 genes is at most 0.42 (left panel).</p>
<p>The dashed gray curve is obtained by the parametric Simes method introduced by <span class="citation">Goeman and Solari (2011)</span>. The comparison between the two curves illustrates the gain in power obtained by using permutations methods to adapt to the dependence between genes. In this example, the parametric Simes method gives the following guarantees: among the 2000 most significant genes, the number of truly differentially expressed genes is at least 928 (right panel). Equivalently, the FDP among these 2000 genes is at most 0.54 (left panel).</p>
<div class="section level3">
<h3 id="differentially-expressed-genes">Differentially expressed genes<a class="anchor" aria-label="anchor" href="#differentially-expressed-genes"></a>
</h3>
<p>In this section we show how to the above curves may be used to address the question:</p>
<blockquote>
<p>Which genes are differentially expressed with high probability?</p>
</blockquote>
<p>To do so, we <em>define</em> differentially expressed genes as the largest set of genes for which the FDP bound is less than a user-given value, for example <span class="math inline">\(q = 0.1\)</span>. This corresponds to drawing a horizontal line in the preceding plot:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.1</span> <span class="co"># FDP budget (user-defined)</span></span>
<span><span class="va">FDP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">resList</span>, <span class="va">predict</span>, what <span class="op">=</span> <span class="st">"FDP"</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">n_DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">FDP</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&lt;=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotConfCurve.html">plotConfCurve</a></span><span class="op">(</span><span class="va">FDP</span>, xmax <span class="op">=</span> <span class="fl">2.5</span><span class="op">*</span><span class="va">n_DEG</span>, col <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">q</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, size <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">n_DEG</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, col <span class="op">=</span> <span class="va">cols</span>, size <span class="op">=</span> <span class="va">size</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">size</span><span class="op">)</span></span>
<span><span class="co"># ggsave(p, file = "conf-curve_DEG.pdf", width = 6, height = 4)</span></span></code></pre></div>
<p>Using <span class="math inline">\(q = 0.1\)</span>, we obtain 940 differentially expressed genes. Note that this gene list has a clear statistical interpretation: with probability <span class="math inline">\(1-\alpha = 0.9\)</span>, the proportion of false positives (that is, genes that are called DE by mistake) is less than <span class="math inline">\(q = 0.1\)</span>.</p>
<p>The above example also illustrates the increase in power obtained by calibration, since the parametric Simes method yields a subset of “only” 733 genes called differentially expressed (with identical statistical guarantees).</p>
</div>
<div class="section level3">
<h3 id="comparison-to-fdr-control">Comparison to FDR control<a class="anchor" aria-label="anchor" href="#comparison-to-fdr-control"></a>
</h3>
<p>We argue that ensuring <em>with high probability</em> that the FDP is less than some user-defined budget, as done above, is more interpretable than the widely used False Discovery Rate (FDR) control introduced by <span class="citation">Benjamini and Hochberg (1995)</span>. Indeed, FDR control guarantees that the <em>expected</em> (or average) FDP among DE genes is less than some user-defined FDR budget. However, <strong>the true FDP may well be far from its expected value</strong>, especially since the variability of the FDP is known to increase with the degree of dependence between genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adjp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="fu"><a href="../reference/all-generics.html">pValues</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">FDR_level</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">n_BH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">adjp</span> <span class="op">&lt;</span> <span class="va">FDR_level</span><span class="op">)</span></span>
<span><span class="va">n_BH</span></span>
<span><span class="co">#&gt; [1] 1713</span></span></code></pre></div>
<p>In our case, the BH procedure would consider 1713 genes as differentially expressed at the FDR level 0.05. This set of DE genes is larger than the 940 genes called DE by our method, but the latter comes with stronger and more interpretable statistical guarantees than FDR control.</p>
</div>
</div>
<div class="section level2">
<h2 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h2>
<!-- For an interactive volcano plot, see the [volcano plot shiny application]( https://shiny-iidea-sanssouci.apps.math.cnrs.fr/). -->
<p>A classical practice in gene expression studies is to define DE genes as those passing both a significance filter (small <span class="math inline">\(p\)</span>-value) and an effect size or “fold change” filter. Here, the fold change of a gene is defined as the difference between the expression medians (on the log scale) of the two groups compared. This double selection by <span class="math inline">\(p\)</span>-value and fold change corresponds to two sets of genes, with positive/negative fold change, which can be represented in the following plot:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, p <span class="op">=</span> <span class="fl">1e-4</span>, r <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_RNAseq_files/figure-html/volcano-1.png" width="768"></p>
<p>This type of plot is called a “volcano plot” <span class="citation">Cui and Churchill (2003)</span>. Post hoc inference makes it possible to obtain statistical guarantees on selections such as the ones represented in the above figure.</p>
<div class="section level3">
<h3 id="custom-statistics-example-using-limma-voom">Custom statistics: example using limma-voom<a class="anchor" aria-label="anchor" href="#custom-statistics-example-using-limma-voom"></a>
</h3>
<p>Post hoc bounds can be calculated for any gene selection. In particular, even if Wilcoxon tests have been performed <em>for the calibration of the post hoc bounds</em>, it is possible to rely on other statistics to <em>select genes of interest</em>. In this section, we illustrate this idea by making a volcano plot based on the <span class="math inline">\(p\)</span>-values and log-fold changes obtained from the limma-voom method of <span class="citation">Law et al. (2014)</span> implemented in the <code>limma</code> package (<span class="citation">Ritchie et al. (2015)</span>). The next code chunk is adapted from the vignette of the <code>limma</code> package:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="co">#&gt; Repeated column names found in count matrix</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="va">Grp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span></span>
<span><span class="va">mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">Grp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">d</span>, <span class="va">mm</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">mm</span><span class="op">)</span></span>
<span><span class="va">contr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span><span class="va">Grp1</span> <span class="op">-</span> <span class="va">Grp0</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">res_lm</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">res_lm</span>, <span class="va">contr</span><span class="op">)</span></span>
<span><span class="va">res_eb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">res_fit</span><span class="op">)</span></span>
<span><span class="va">TT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">res_eb</span>, sort.by <span class="op">=</span> <span class="st">"none"</span>, number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<p>The next plot suggests that the <span class="math inline">\(p\)</span>-values obtained from limma-voom and from Wilcoxon tests are consistent.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>wilcox <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="../reference/all-generics.html">pValues</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span>, limma <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">TT</span><span class="op">$</span><span class="va">P.Value</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">limma</span>, y <span class="op">=</span> <span class="va">wilcox</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#10101010"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"p-values (log-scale)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_RNAseq_files/figure-html/compare-p-values-1.png" width="768"></p>
<p>Finally we obtain the volcano plot based on limma statistics.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/volcanoPlot.html">volcanoPlot</a></span><span class="op">(</span><span class="va">res</span>, </span>
<span>            fold_changes <span class="op">=</span> <span class="va">TT</span><span class="op">$</span><span class="va">logFC</span>, </span>
<span>            p_values <span class="op">=</span> <span class="va">TT</span><span class="op">$</span><span class="va">P.Value</span>, </span>
<span>            p <span class="op">=</span> <span class="fl">1e-4</span>, r <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="post-hoc_differential-expression_RNAseq_files/figure-html/volcano-limma-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] edgeR_3.38.4       limma_3.52.2       matrixStats_0.62.0 sanssouci_0.12.7  </span></span>
<span><span class="co">#&gt; [5] ggplot2_3.3.6     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] matrixTests_0.1.9.1 locfit_1.5-9.6      xfun_0.32          </span></span>
<span><span class="co">#&gt;  [4] bslib_0.4.0         purrr_0.3.4         lattice_0.20-45    </span></span>
<span><span class="co">#&gt;  [7] colorspace_2.0-3    vctrs_0.4.1         generics_0.1.3     </span></span>
<span><span class="co">#&gt; [10] htmltools_0.5.3     yaml_2.3.5          utf8_1.2.2         </span></span>
<span><span class="co">#&gt; [13] rlang_1.0.4         pkgdown_2.0.6       jquerylib_0.1.4    </span></span>
<span><span class="co">#&gt; [16] pillar_1.8.1        glue_1.6.2          withr_2.5.0        </span></span>
<span><span class="co">#&gt; [19] lifecycle_1.0.1     stringr_1.4.1       munsell_0.5.0      </span></span>
<span><span class="co">#&gt; [22] gtable_0.3.0        ragg_1.2.2          codetools_0.2-18   </span></span>
<span><span class="co">#&gt; [25] memoise_2.0.1       evaluate_0.16       labeling_0.4.2     </span></span>
<span><span class="co">#&gt; [28] knitr_1.40          fastmap_1.1.0       fansi_1.0.3        </span></span>
<span><span class="co">#&gt; [31] highr_0.9           Rcpp_1.0.9          scales_1.2.1       </span></span>
<span><span class="co">#&gt; [34] cachem_1.0.6        desc_1.4.1          jsonlite_1.8.0     </span></span>
<span><span class="co">#&gt; [37] farver_2.1.1        systemfonts_1.0.4   fs_1.5.2           </span></span>
<span><span class="co">#&gt; [40] textshaping_0.3.6   digest_0.6.29       stringi_1.7.8      </span></span>
<span><span class="co">#&gt; [43] grid_4.2.1          rprojroot_2.0.3     cli_3.3.0          </span></span>
<span><span class="co">#&gt; [46] tools_4.2.1         magrittr_2.0.3      sass_0.4.2         </span></span>
<span><span class="co">#&gt; [49] tibble_3.1.8        pkgconfig_2.0.3     Matrix_1.4-1       </span></span>
<span><span class="co">#&gt; [52] rmarkdown_2.16      R6_2.5.1            compiler_4.2.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>To re-build this vignette from its source, use:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_differential-expression_RNAseq.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"pdf_document"</span><span class="op">)</span></span>
<span><span class="co"># To keep intermediate files, add option 'clean = FALSE'</span></span>
<span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"post-hoc_differential-expression_RNAseq.Rmd"</span>, output_format <span class="op">=</span> <span class="st">"html_document"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-benjamini95controlling">
<p>Benjamini, Yoav, and Yosef Hochberg. 1995. “Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing.” <em>Journal of the Royal Statistical Society: Series B (Methodological)</em> 57 (1): 289–300.</p>
</div>
<div id="ref-blanchard20post-hoc">
<p>Blanchard, Gilles, Pierre Neuvial, and Etienne Roquain. 2020. “Post Hoc Confidence Bounds on False Positives Using Reference Families.” <em>Annals of Statistics</em> 48 (3): 1281–1303. <a href="https://projecteuclid.org/euclid.aos/1594972818" class="external-link">https://projecteuclid.org/euclid.aos/1594972818</a>.</p>
</div>
<div id="ref-BNR:chap">
<p>———. 2021. “On agnostic post hoc approaches to false positive control.” In <em>Handbook of Multiple Comparisons</em>, edited by Xinping Cui, Thorsten Dickhaus, Ying Ding, and Jason C. Hsu. Handbooks of Modern Statistical Methods. Chapman<br>
&amp; Hall/CRC. <a href="https://hal.archives-ouvertes.fr/hal-02320543" class="external-link">https://hal.archives-ouvertes.fr/hal-02320543</a>.</p>
</div>
<div id="ref-tcga2014blca">
<p>Cancer Genome Atlas Research Network, and others. 2014. “Comprehensive Molecular Characterization of Urothelial Bladder Carcinoma.” <em>Nature</em> 507 (7492): 315.</p>
</div>
<div id="ref-CC2003">
<p>Cui, Xiangqin, and Gary A Churchill. 2003. “Statistical Tests for Differential Expression in cDNA Microarray Experiments.” <em>Genome Biol</em> 4 (4): 210.</p>
</div>
<div id="ref-GS2011">
<p>Goeman, Jelle J, and Aldo Solari. 2011. “Multiple Testing for Exploratory Research.” <em>Statistical Science</em> 26 (4): 584–97.</p>
</div>
<div id="ref-law2014voom">
<p>Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom: Precision Weights Unlock Linear Model Analysis Tools for Rna-Seq Read Counts.” <em>Genome Biology</em> 15 (2): 1–17.</p>
</div>
<div id="ref-limma">
<p>Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. “limma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.” <em>Nucleic Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.math.univ-toulouse.fr/~pneuvial/" class="external-link">Pierre Neuvial</a>, <a href="https://www.imo.universite-paris-saclay.fr/~blanchard/" class="external-link">Gilles Blanchard</a>, <a href="https://durandg12.github.io/" class="external-link">Guillermo Durand</a>, Nicolas Enjalbert-Courrech, <a href="https://etienneroquain-81.webself.net/" class="external-link">Etienne Roquain</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
